

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>INDRA Statements (indra.statements) &mdash; INDRA 1.5.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="INDRA 1.5.0 documentation" href="../index.html"/>
        <link rel="up" title="INDRA modules reference" href="index.html"/>
        <link rel="next" title="Processors for model input (indra.sources)" href="sources/index.html"/>
        <link rel="prev" title="INDRA modules reference" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="../index.html" class="icon icon-home"> INDRA
        

        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#installing-python">Installing Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#installing-indra">Installing INDRA</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#installing-via-github">Installing via Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#cloning-the-source-code-from-github">Cloning the source code from Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#installing-releases-with-pip">Installing releases with pip</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../installation.html#indra-dependencies">INDRA dependencies</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#pysb-and-bionetgen">PySB and BioNetGen</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#pyjnius">Pyjnius</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#graphviz">Graphviz</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#matplotlib">Matplotlib</a></li>
<li class="toctree-l3"><a class="reference internal" href="../installation.html#optional-additional-dependencies">Optional additional dependencies</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../getting_started.html">Getting started with INDRA</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../getting_started.html#importing-indra-and-its-modules">Importing INDRA and its modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="../getting_started.html#basic-usage-examples">Basic usage examples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#reading-a-sentence-with-trips">Reading a sentence with TRIPS</a></li>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#reading-a-pubmed-central-article-with-reach">Reading a PubMed Central article with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#getting-the-neighborhood-of-proteins-from-the-bel-large-corpus">Getting the neighborhood of proteins from the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#getting-paths-between-two-proteins-from-pathwaycommons-biopax">Getting paths between two proteins from PathwayCommons (BioPAX)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#constructing-indra-statements-manually">Constructing INDRA Statements manually</a></li>
<li class="toctree-l3"><a class="reference internal" href="../getting_started.html#assembling-a-pysb-model-and-exporting-to-sbml">Assembling a PySB model and exporting to SBML</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="index.html">INDRA modules reference</a><ul class="current">
<li class="toctree-l2 current"><a class="current reference internal" href="#">INDRA Statements (<code class="docutils literal"><span class="pre">indra.statements</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="sources/index.html">Processors for model input (<code class="docutils literal"><span class="pre">indra.sources</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="sources/bel/index.html">BEL (<code class="docutils literal"><span class="pre">indra.sources.bel</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="sources/bel/index.html#module-indra.sources.bel.bel_api">BEL API (<code class="docutils literal"><span class="pre">indra.sources.bel.bel_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="sources/bel/index.html#module-indra.sources.bel.processor">BEL Processor (<code class="docutils literal"><span class="pre">indra.sources.bel.processor</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="sources/biopax/index.html">Biopax (<code class="docutils literal"><span class="pre">indra.sources.biopax</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="sources/biopax/index.html#module-indra.sources.biopax.biopax_api">Biopax API (<code class="docutils literal"><span class="pre">indra.sources.biopax.biopax_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="sources/biopax/index.html#module-indra.sources.biopax.processor">Biopax Processor (<code class="docutils literal"><span class="pre">indra.sources.biopax.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="sources/biopax/index.html#module-indra.sources.biopax.pathway_commons_client">Pathway Commons Client (<code class="docutils literal"><span class="pre">indra.sources.biopax.pathway_commons_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="sources/reach/index.html">REACH (<code class="docutils literal"><span class="pre">indra.sources.reach</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="sources/reach/index.html#module-indra.sources.reach.reach_api">REACH API (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="sources/reach/index.html#module-indra.sources.reach.processor">REACH Processor (<code class="docutils literal"><span class="pre">indra.sources.reach.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="sources/reach/index.html#module-indra.sources.reach.reach_reader">REACH reader (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_reader</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="sources/trips/index.html">TRIPS (<code class="docutils literal"><span class="pre">indra.sources.trips</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="sources/trips/index.html#module-indra.sources.trips.trips_api">TRIPS API (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="sources/trips/index.html#module-indra.sources.trips.processor">TRIPS Processor (<code class="docutils literal"><span class="pre">indra.sources.trips.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="sources/trips/index.html#module-indra.sources.trips.trips_client">TRIPS Client (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="sources/ndex_cx/index.html">NDEx CX API (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx_api</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="sources/ndex_cx/index.html#ndex-cx-processor-indra-sources-ndex-cx-processor">NDEx CX Processor (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx.processor</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="databases/index.html">Database clients (<code class="docutils literal"><span class="pre">indra.databases</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.hgnc_client">HGNC client (<code class="docutils literal"><span class="pre">indra.hgnc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.uniprot_client">Uniprot client (<code class="docutils literal"><span class="pre">indra.databases.uniprot_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.chebi_client">ChEBI client (<code class="docutils literal"><span class="pre">indra.databases.chebi_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.biogrid_client">BioGRID client (<code class="docutils literal"><span class="pre">indra.databases.biogrid_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.context_client">Cell type context client (<code class="docutils literal"><span class="pre">indra.databases.context_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.relevance_client">Network relevance client (<code class="docutils literal"><span class="pre">indra.databases.relevance_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.ndex_client">NDEx client (<code class="docutils literal"><span class="pre">indra.databases.ndex_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="databases/index.html#module-indra.databases.cbio_client">cBio portal client (<code class="docutils literal"><span class="pre">indra.databases.cbio_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="literature/index.html">Literature clients (<code class="docutils literal"><span class="pre">indra.literature</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="literature/index.html#module-indra.literature.pubmed_client">Pubmed client (<code class="docutils literal"><span class="pre">indra.literature.pubmed_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="literature/index.html#module-indra.literature.pmc_client">Pubmed Central client (<code class="docutils literal"><span class="pre">indra.literature.pmc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="literature/index.html#module-indra.literature.crossref_client">CrossRef client (<code class="docutils literal"><span class="pre">indra.literature.crossref_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="literature/index.html#module-indra.literature.elsevier_client">Elsevier client (<code class="docutils literal"><span class="pre">indra.literature.elsevier_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="preassembler/index.html">Preassembly (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="preassembler/index.html#module-indra.preassembler">Preassembler (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="preassembler/index.html#module-indra.preassembler.grounding_mapper">Entity grounding curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.grounding_mapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="preassembler/index.html#module-indra.preassembler.sitemapper">Site curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.sitemapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="preassembler/index.html#module-indra.preassembler.hierarchy_manager">Hierarchy manager (<code class="docutils literal"><span class="pre">indra.preassembler.hierarchy_manager</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="belief/index.html">Belief Engine (<code class="docutils literal"><span class="pre">indra.belief</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="mechlinker/index.html">Mechanism Linker (<code class="docutils literal"><span class="pre">indra.mechlinker</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="assemblers/index.html">Assemblers of model output (<code class="docutils literal"><span class="pre">indra.assemblers</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="assemblers/pysb_assembler.html">Executable PySB models (<code class="docutils literal"><span class="pre">indra.assemblers.pysb_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="assemblers/cx_assembler.html">Cytoscape networks (<code class="docutils literal"><span class="pre">indra.assemblers.cx_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="assemblers/english_assembler.html">Natural language (<code class="docutils literal"><span class="pre">indra.assemblers.english_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="assemblers/graph_assembler.html">Node-edge graphs (<code class="docutils literal"><span class="pre">indra.assemblers.graph_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="assemblers/sif_assembler.html">SIF / Boolean networks (<code class="docutils literal"><span class="pre">indra.assemblers.sif_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="assemblers/index_card_assembler.html">MITRE &#8220;index cards&#8221; (<code class="docutils literal"><span class="pre">indra.assemblers.index_card_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="assemblers/sbgn_assembler.html">SBGN output (<code class="docutils literal"><span class="pre">indra.assemblers.sbgn_assembler</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="explanation/index.html">Explanation (<code class="docutils literal"><span class="pre">indra.explanation</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="explanation/index.html#module-indra.explanation.model_checker">Check whether a rule-based model satisfies a property (<code class="docutils literal"><span class="pre">indra.explanation.model_checker</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="tools/index.html">Tools (<code class="docutils literal"><span class="pre">indra.tools</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="tools/index.html#module-indra.tools.assemble_corpus">Run assembly components in a pipeline (<code class="docutils literal"><span class="pre">indra.tools.assemble_corpus</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="tools/index.html#module-indra.tools.gene_network">Build a network from a gene list (<code class="docutils literal"><span class="pre">indra.tools.gene_network</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="tools/index.html#module-indra.tools.executable_subnetwork">Build an executable model from a fragment of a large network (<code class="docutils literal"><span class="pre">indra.tools.executable_subnetwork</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="tools/index.html#module-indra.tools.incremental_model">Build a model incrementally over time (<code class="docutils literal"><span class="pre">indra.tools.incremental_model</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../tutorials/index.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/nl_modeling.html">Using natural language to build models</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/nl_modeling.html#read-indra-statements-from-a-natural-language-string">Read INDRA Statements from a natural language string</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/nl_modeling.html#assemble-the-indra-statements-into-a-rule-based-executable-model">Assemble the INDRA Statements into a rule-based executable model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/nl_modeling.html#exporting-the-model-into-other-common-formats">Exporting the model into other common formats</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/machine_reading.html">Large-Scale Machine Reading with Starcluster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading.html#install-reach">Install REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading.html#install-amazon-s3-support">Install Amazon S3 support</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading.html#install-other-dependencies">Install other dependencies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading.html#assemble-a-corpus-of-pmids">Assemble a Corpus of PMIDs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading.html#process-the-papers-with-reach">Process the papers with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading.html#extract-indra-statements-from-the-reach-output-on-s3">Extract INDRA Statements from the REACH output on S3</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading.html#running-the-whole-pipeline-with-one-script">Running the whole pipeline with one script</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/machine_reading_aws.html">Large-Scale Machine Reading with Amazon Batch</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/machine_reading_aws.html#how-it-works">How it Works</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../tutorials/gene_network.html">Assembling everything known about a particular gene</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/gene_network.html#collect-mechanisms-from-pathwaycommons-and-the-bel-large-corpus">Collect mechanisms from PathwayCommons and the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/gene_network.html#collect-a-list-of-publications-that-discuss-the-gene-of-interest">Collect a list of publications that discuss the gene of interest</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/gene_network.html#get-the-full-text-or-abstract-corresponding-to-the-publications">Get the full text or abstract corresponding to the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/gene_network.html#read-the-content-of-the-publications">Read the content of the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/gene_network.html#combine-all-statements-and-run-pre-assembly">Combine all statements and run pre-assembly</a></li>
<li class="toctree-l3"><a class="reference internal" href="../tutorials/gene_network.html#assemble-the-statements-into-a-network-model">Assemble the statements into a network model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../rest_api.html">REST API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../rest_api.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rest_api.html#launching-the-rest-service">Launching the REST service</a></li>
<li class="toctree-l2"><a class="reference internal" href="../rest_api.html#documentation">Documentation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../index.html">INDRA</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../index.html">Docs</a> &raquo;</li>
      
          <li><a href="index.html">INDRA modules reference</a> &raquo;</li>
      
    <li>INDRA Statements (<code class="docutils literal"><span class="pre">indra.statements</span></code>)</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../_sources/modules/statements.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="module-indra.statements">
<span id="indra-statements-indra-statements"></span><h1>INDRA Statements (<a class="reference internal" href="#module-indra.statements" title="indra.statements"><code class="xref py py-mod docutils literal"><span class="pre">indra.statements</span></code></a>)<a class="headerlink" href="#module-indra.statements" title="Permalink to this headline">¶</a></h1>
<p>Statements represent mechanistic relationships between biological agents.</p>
<p>Statement classes follow an inheritance hierarchy, with all Statement types
inheriting from the parent class <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">Statement</span></code></a>. At
the next level in the hierarchy are the following classes:</p>
<ul class="simple">
<li><a class="reference internal" href="#indra.statements.Complex" title="indra.statements.Complex"><code class="xref py py-class docutils literal"><span class="pre">Complex</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Modification" title="indra.statements.Modification"><code class="xref py py-class docutils literal"><span class="pre">Modification</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.SelfModification" title="indra.statements.SelfModification"><code class="xref py py-class docutils literal"><span class="pre">SelfModification</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.RegulateActivity" title="indra.statements.RegulateActivity"><code class="xref py py-class docutils literal"><span class="pre">RegulateActivity</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.RegulateAmount" title="indra.statements.RegulateAmount"><code class="xref py py-class docutils literal"><span class="pre">RegulateAmount</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.ActiveForm" title="indra.statements.ActiveForm"><code class="xref py py-class docutils literal"><span class="pre">ActiveForm</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Translocation" title="indra.statements.Translocation"><code class="xref py py-class docutils literal"><span class="pre">Translocation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Gef" title="indra.statements.Gef"><code class="xref py py-class docutils literal"><span class="pre">Gef</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Gap" title="indra.statements.Gap"><code class="xref py py-class docutils literal"><span class="pre">Gap</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Conversion" title="indra.statements.Conversion"><code class="xref py py-class docutils literal"><span class="pre">Conversion</span></code></a></li>
</ul>
<p>There are several types of Statements representing post-translational
modifications that further inherit from
<a class="reference internal" href="#indra.statements.Modification" title="indra.statements.Modification"><code class="xref py py-class docutils literal"><span class="pre">Modification</span></code></a>:</p>
<ul class="simple">
<li><a class="reference internal" href="#indra.statements.Phosphorylation" title="indra.statements.Phosphorylation"><code class="xref py py-class docutils literal"><span class="pre">Phosphorylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Dephosphorylation" title="indra.statements.Dephosphorylation"><code class="xref py py-class docutils literal"><span class="pre">Dephosphorylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Ubiquitination" title="indra.statements.Ubiquitination"><code class="xref py py-class docutils literal"><span class="pre">Ubiquitination</span></code></a></li>
<li><code class="xref py py-class docutils literal"><span class="pre">Debiquitination</span></code></li>
<li><a class="reference internal" href="#indra.statements.Sumoylation" title="indra.statements.Sumoylation"><code class="xref py py-class docutils literal"><span class="pre">Sumoylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Desumoylation" title="indra.statements.Desumoylation"><code class="xref py py-class docutils literal"><span class="pre">Desumoylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Hydroxylation" title="indra.statements.Hydroxylation"><code class="xref py py-class docutils literal"><span class="pre">Hydroxylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Dehydroxylation" title="indra.statements.Dehydroxylation"><code class="xref py py-class docutils literal"><span class="pre">Dehydroxylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Acetylation" title="indra.statements.Acetylation"><code class="xref py py-class docutils literal"><span class="pre">Acetylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Deacetylation" title="indra.statements.Deacetylation"><code class="xref py py-class docutils literal"><span class="pre">Deacetylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Glycosylation" title="indra.statements.Glycosylation"><code class="xref py py-class docutils literal"><span class="pre">Glycosylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Deglycosylation" title="indra.statements.Deglycosylation"><code class="xref py py-class docutils literal"><span class="pre">Deglycosylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Farnesylation" title="indra.statements.Farnesylation"><code class="xref py py-class docutils literal"><span class="pre">Farnesylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Defarnesylation" title="indra.statements.Defarnesylation"><code class="xref py py-class docutils literal"><span class="pre">Defarnesylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Geranylgeranylation" title="indra.statements.Geranylgeranylation"><code class="xref py py-class docutils literal"><span class="pre">Geranylgeranylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Degeranylgeranylation" title="indra.statements.Degeranylgeranylation"><code class="xref py py-class docutils literal"><span class="pre">Degeranylgeranylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Palmitoylation" title="indra.statements.Palmitoylation"><code class="xref py py-class docutils literal"><span class="pre">Palmitoylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Depalmitoylation" title="indra.statements.Depalmitoylation"><code class="xref py py-class docutils literal"><span class="pre">Depalmitoylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Myristoylation" title="indra.statements.Myristoylation"><code class="xref py py-class docutils literal"><span class="pre">Myristoylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Demyristoylation" title="indra.statements.Demyristoylation"><code class="xref py py-class docutils literal"><span class="pre">Demyristoylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Ribosylation" title="indra.statements.Ribosylation"><code class="xref py py-class docutils literal"><span class="pre">Ribosylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Deribosylation" title="indra.statements.Deribosylation"><code class="xref py py-class docutils literal"><span class="pre">Deribosylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Methylation" title="indra.statements.Methylation"><code class="xref py py-class docutils literal"><span class="pre">Methylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Demethylation" title="indra.statements.Demethylation"><code class="xref py py-class docutils literal"><span class="pre">Demethylation</span></code></a></li>
</ul>
<p>There are additional subtypes of <a class="reference internal" href="#indra.statements.SelfModification" title="indra.statements.SelfModification"><code class="xref py py-class docutils literal"><span class="pre">SelfModification</span></code></a>:</p>
<ul class="simple">
<li><a class="reference internal" href="#indra.statements.Autophosphorylation" title="indra.statements.Autophosphorylation"><code class="xref py py-class docutils literal"><span class="pre">Autophosphorylation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Transphosphorylation" title="indra.statements.Transphosphorylation"><code class="xref py py-class docutils literal"><span class="pre">Transphosphorylation</span></code></a></li>
</ul>
<p>Interactions between proteins are often described simply in terms of their
effect on a protein&#8217;s &#8220;activity&#8221;, e.g., &#8220;Active MEK activates ERK&#8221;, or &#8220;DUSP6
inactives ERK&#8221;.  These types of relationships are indicated by the
<a class="reference internal" href="#indra.statements.RegulateActivity" title="indra.statements.RegulateActivity"><code class="xref py py-class docutils literal"><span class="pre">RegulateActivity</span></code></a> abstract base class which has subtypes</p>
<ul class="simple">
<li><a class="reference internal" href="#indra.statements.Activation" title="indra.statements.Activation"><code class="xref py py-class docutils literal"><span class="pre">Activation</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.Inhibition" title="indra.statements.Inhibition"><code class="xref py py-class docutils literal"><span class="pre">Inhibition</span></code></a></li>
</ul>
<p>while the <a class="reference internal" href="#indra.statements.RegulateAmount" title="indra.statements.RegulateAmount"><code class="xref py py-class docutils literal"><span class="pre">RegulateAmount</span></code></a> abstract base class has subtypes</p>
<ul class="simple">
<li><a class="reference internal" href="#indra.statements.IncreaseAmount" title="indra.statements.IncreaseAmount"><code class="xref py py-class docutils literal"><span class="pre">IncreaseAmount</span></code></a></li>
<li><a class="reference internal" href="#indra.statements.DecreaseAmount" title="indra.statements.DecreaseAmount"><code class="xref py py-class docutils literal"><span class="pre">DecreaseAmount</span></code></a></li>
</ul>
<p>Statements involve one or more biological <em>Agents</em>, typically proteins,
represented by the class <a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>. Agents can have several types
of context specified on them including</p>
<ul class="simple">
<li>a specific post-translational modification state (indicated by one or
more instances of <a class="reference internal" href="#indra.statements.ModCondition" title="indra.statements.ModCondition"><code class="xref py py-class docutils literal"><span class="pre">ModCondition</span></code></a>),</li>
<li>other bound Agents (<a class="reference internal" href="#indra.statements.BoundCondition" title="indra.statements.BoundCondition"><code class="xref py py-class docutils literal"><span class="pre">BoundCondition</span></code></a>),</li>
<li>mutations (<a class="reference internal" href="#indra.statements.MutCondition" title="indra.statements.MutCondition"><code class="xref py py-class docutils literal"><span class="pre">MutCondition</span></code></a>),</li>
<li>an activity state (<a class="reference internal" href="#indra.statements.ActivityCondition" title="indra.statements.ActivityCondition"><code class="xref py py-class docutils literal"><span class="pre">ActivityCondition</span></code></a>), and</li>
<li>cellular location</li>
</ul>
<p>The <em>active</em> form of an agent (in terms of its post-translational modifications
or bound state) is indicated by an instance of the class
<a class="reference internal" href="#indra.statements.ActiveForm" title="indra.statements.ActiveForm"><code class="xref py py-class docutils literal"><span class="pre">ActiveForm</span></code></a>.</p>
<p>Agents also carry grounding information which links them to database entries.
These database references are represented as a dictionary in the <cite>db_refs</cite>
attribute of each Agent. The dictionary can have multiple entries. For
instance, INDRA&#8217;s input Processors produce genes and proteins that carry both
UniProt and HGNC IDs in db_refs, whenever possible. Bioentities provides a name
space for protein families that are typically used in the literature.  More
information about Bioentities can be found here:
<a class="reference external" href="https://github.com/sorgerlab/bioentities">https://github.com/sorgerlab/bioentities</a></p>
<table border="1" class="docutils">
<colgroup>
<col width="35%" />
<col width="26%" />
<col width="38%" />
</colgroup>
<thead valign="bottom">
<tr class="row-odd"><th class="head">Type</th>
<th class="head">Database</th>
<th class="head">Example</th>
</tr>
</thead>
<tbody valign="top">
<tr class="row-even"><td>Gene/Protein</td>
<td>HGNC</td>
<td>{&#8216;HGNC&#8217;: &#8216;11998&#8217;}</td>
</tr>
<tr class="row-odd"><td>Gene/Protein</td>
<td>UniProt</td>
<td>{&#8216;UP&#8217;: &#8216;P04637&#8217;}</td>
</tr>
<tr class="row-even"><td>Gene/Protein family</td>
<td>Bioentities</td>
<td>{&#8216;BE&#8217;: &#8216;ERK&#8217;}</td>
</tr>
<tr class="row-odd"><td>Gene/Protein family</td>
<td>InterPro</td>
<td>{&#8216;IP&#8217;: &#8216;IPR000308&#8217;}</td>
</tr>
<tr class="row-even"><td>Gene/Protein family</td>
<td>Pfam</td>
<td>{&#8216;PF&#8217;: &#8216;PF00071&#8217;}</td>
</tr>
<tr class="row-odd"><td>Gene/Protein family</td>
<td>NextProt family</td>
<td>{&#8216;NXPFAM&#8217;: &#8216;03114&#8217;}</td>
</tr>
<tr class="row-even"><td>Chemical</td>
<td>ChEBI</td>
<td>{&#8216;CHEBI&#8217;: &#8216;CHEBI:63637&#8217;}</td>
</tr>
<tr class="row-odd"><td>Chemical</td>
<td>PubChem</td>
<td>{&#8216;PUBCHEM&#8217;: &#8216;42611257&#8217;}</td>
</tr>
<tr class="row-even"><td>Metabolite</td>
<td>HMDB</td>
<td>{&#8216;HMDB&#8217;: &#8216;HMDB00122&#8217;}</td>
</tr>
<tr class="row-odd"><td>Process, location, etc.</td>
<td>GO</td>
<td>{&#8216;GO&#8217;: &#8216;<a class="reference external" href="GO:0006915">GO:0006915</a>&#8216;}</td>
</tr>
<tr class="row-even"><td>Process, disease, etc.</td>
<td>MeSH</td>
<td>{&#8216;MESH&#8217;: &#8216;D008113&#8217;}</td>
</tr>
<tr class="row-odd"><td>General terms</td>
<td>NCIT</td>
<td>{&#8216;NCIT&#8217;: &#8216;C28597&#8217;}</td>
</tr>
<tr class="row-even"><td>Raw text</td>
<td>TEXT</td>
<td>{&#8216;TEXT&#8217;: &#8216;Nf-kappaB&#8217;}</td>
</tr>
</tbody>
</table>
<p>The evidence for a given Statement, which could include relevant citations,
database identifiers, and passages of text from the scientific literature, is
contained in one or more <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a> objects associated with the
Statement.</p>
<dl class="class">
<dt id="indra.statements.Acetylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Acetylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Acetylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Acetylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Acetylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Activation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Activation</code><span class="sig-paren">(</span><em>subj</em>, <em>obj</em>, <em>obj_activity=u'activity'</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Activation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Activation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.RegulateActivity" title="indra.statements.RegulateActivity"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.RegulateActivity</span></code></a></p>
<p>Indicates that a protein activates another protein.</p>
<p>This statement is intended to be used for physical interactions where the
mechanism of activation is not explicitly specified, which is often the
case for descriptions of mechanisms extracted from the literature.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>subj</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The agent responsible for the change in activity, i.e., the &#8220;upstream&#8221;
node.</li>
<li><strong>obj</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The agent whose activity is influenced by the subject, i.e., the
&#8220;downstream&#8221; node.</li>
<li><strong>obj_activity</strong> (<em>Optional[str]</em>) &#8211; The activity of the obj Agent that is affected, e.g., its &#8220;kinase&#8221;
activity.</li>
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; Evidence objects in support of the modification.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>MEK (MAP2K1) activates the kinase activity of ERK (MAPK1):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mek</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAP2K1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">erk</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAPK1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">act</span> <span class="o">=</span> <span class="n">Activation</span><span class="p">(</span><span class="n">mek</span><span class="p">,</span> <span class="n">erk</span><span class="p">,</span> <span class="s1">&#39;kinase&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.ActiveForm">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">ActiveForm</code><span class="sig-paren">(</span><em>agent</em>, <em>activity</em>, <em>is_active</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#ActiveForm"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.ActiveForm" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Specifies conditions causing an Agent to be active or inactive.</p>
<p>Types of conditions influencing a specific type of biochemical activity can
include modifications, bound Agents, and mutations.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>agent</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The Agent in a particular active or inactive state. The sets
of ModConditions, BoundConditions, and MutConditions on the given
Agent instance indicate the relevant conditions.</li>
<li><strong>activity</strong> (<em>str</em>) &#8211; The type of activity influenced by the given set of conditions,
e.g., &#8220;kinase&#8221;.</li>
<li><strong>is_active</strong> (<em>bool</em>) &#8211; Whether the conditions are activating (True) or inactivating (False).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.ActivityCondition">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">ActivityCondition</code><span class="sig-paren">(</span><em>activity_type</em>, <em>is_active</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#ActivityCondition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.ActivityCondition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>An active or inactive state of a protein.</p>
<p class="rubric">Examples</p>
<p>Kinase-active MAP2K1:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mek_active</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAP2K1&#39;</span><span class="p">,</span>
<span class="gp">... </span>                   <span class="n">activity</span><span class="o">=</span><span class="n">ActivityCondition</span><span class="p">(</span><span class="s1">&#39;kinase&#39;</span><span class="p">,</span> <span class="kc">True</span><span class="p">))</span>
</pre></div>
</div>
<p>Transcriptionally inactive FOXO3:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">foxo_inactive</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;FOXO3&#39;</span><span class="p">,</span>
<span class="gp">... </span>                    <span class="n">activity</span><span class="o">=</span><span class="n">ActivityCondition</span><span class="p">(</span><span class="s1">&#39;transcription&#39;</span><span class="p">,</span> <span class="kc">False</span><span class="p">))</span>
</pre></div>
</div>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>activity_type</strong> (<em>str</em>) &#8211; The type of activity, e.g. &#8216;kinase&#8217;. The basic, unspecified molecular
activity is represented as &#8216;activity&#8217;. Examples of other activity
types are &#8216;kinase&#8217;, &#8216;phosphatase&#8217;, &#8216;catalytic&#8217;, &#8216;transcription&#8217;,
etc.</li>
<li><strong>is_active</strong> (<em>bool</em>) &#8211; Specifies whether the given activity type is present or absent.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Agent">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Agent</code><span class="sig-paren">(</span><em>name</em>, <em>mods=None</em>, <em>activity=None</em>, <em>bound_conditions=None</em>, <em>mutations=None</em>, <em>location=None</em>, <em>db_refs=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Agent"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Agent" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>A molecular entity, e.g., a protein.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>name</strong> (<em>str</em>) &#8211; The name of the agent, preferably a canonicalized name such as an
HGNC gene name.</li>
<li><strong>mods</strong> (list of <a class="reference internal" href="#indra.statements.ModCondition" title="indra.statements.ModCondition"><code class="xref py py-class docutils literal"><span class="pre">ModCondition</span></code></a>) &#8211; Modification state of the agent.</li>
<li><strong>bound_conditions</strong> (list of <a class="reference internal" href="#indra.statements.BoundCondition" title="indra.statements.BoundCondition"><code class="xref py py-class docutils literal"><span class="pre">BoundCondition</span></code></a>) &#8211; Other agents bound to the agent in this context.</li>
<li><strong>mutations</strong> (list of <a class="reference internal" href="#indra.statements.MutCondition" title="indra.statements.MutCondition"><code class="xref py py-class docutils literal"><span class="pre">MutCondition</span></code></a>) &#8211; Amino acid mutations of the agent.</li>
<li><strong>activity</strong> (<a class="reference internal" href="#indra.statements.ActivityCondition" title="indra.statements.ActivityCondition"><code class="xref py py-class docutils literal"><span class="pre">ActivityCondition</span></code></a>) &#8211; Activity of the agent.</li>
<li><strong>location</strong> (<em>str</em>) &#8211; Cellular location of the agent. Must be a valid name (e.g. &#8220;nucleus&#8221;)
or identifier (e.g. &#8220;<a class="reference external" href="GO:0005634">GO:0005634</a>&#8221;)for a GO cellular compartment.</li>
<li><strong>db_refs</strong> (<em>dict</em>) &#8211; Dictionary of database identifiers associated with this agent.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Autophosphorylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Autophosphorylation</code><span class="sig-paren">(</span><em>enz</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Autophosphorylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Autophosphorylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.SelfModification" title="indra.statements.SelfModification"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.SelfModification</span></code></a></p>
<p>Intramolecular autophosphorylation, i.e., in <em>cis</em>.</p>
<p class="rubric">Examples</p>
<p>p38 bound to TAB1 cis-autophosphorylates itself (see <span>PMID:</span><a class="reference external" href="http://www.ncbi.nlm.nih.gov/pubmed/19155529">19155529</a>).</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">tab1</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;TAB1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">p38_tab1</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;P38&#39;</span><span class="p">,</span> <span class="n">bound_conditions</span><span class="o">=</span><span class="p">[</span><span class="n">BoundCondition</span><span class="p">(</span><span class="n">tab1</span><span class="p">)])</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">autophos</span> <span class="o">=</span> <span class="n">Autophosphorylation</span><span class="p">(</span><span class="n">p38_tab1</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.BoundCondition">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">BoundCondition</code><span class="sig-paren">(</span><em>agent</em>, <em>is_bound=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#BoundCondition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.BoundCondition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Identify Agents bound (or not bound) to a given Agent in a given context.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>agent</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; Instance of Agent.</li>
<li><strong>is_bound</strong> (<em>bool</em>) &#8211; Specifies whether the given Agent is bound or unbound in the current
context. Default is True.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>EGFR bound to EGF:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">egf</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;EGF&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">egfr</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;EGFR&#39;</span><span class="p">,</span> <span class="n">bound_conditions</span><span class="o">=</span><span class="p">[</span><span class="n">BoundCondition</span><span class="p">(</span><span class="n">egf</span><span class="p">)])</span>
</pre></div>
</div>
<p>BRAF <em>not</em> bound to a 14-3-3 protein (YWHAB):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">ywhab</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;YWHAB&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">braf</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;BRAF&#39;</span><span class="p">,</span> <span class="n">bound_conditions</span><span class="o">=</span><span class="p">[</span><span class="n">BoundCondition</span><span class="p">(</span><span class="n">ywhab</span><span class="p">,</span> <span class="kc">False</span><span class="p">)])</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Complex">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Complex</code><span class="sig-paren">(</span><em>members</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Complex"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Complex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>A set of proteins observed to be in a complex.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>members</strong> (list of <a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The set of proteins in the complex.</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>BRAF is observed to be in a complex with RAF1:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">braf</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;BRAF&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">raf1</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;RAF1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">cplx</span> <span class="o">=</span> <span class="n">Complex</span><span class="p">([</span><span class="n">braf</span><span class="p">,</span> <span class="n">raf1</span><span class="p">])</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Conversion">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Conversion</code><span class="sig-paren">(</span><em>subj</em>, <em>obj_from=None</em>, <em>obj_to=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Conversion"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Conversion" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Conversion of molecular species mediated by a controller protein.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>subj</strong> (<em>:py:class`indra.statement.Agent`</em>) &#8211; The protein mediating the conversion.</li>
<li><strong>obj_from</strong> (list of <code class="xref py py-class docutils literal"><span class="pre">indra.statement.Agent</span></code>) &#8211; The list of molecular species being consumed by the conversion.</li>
<li><strong>obj_to</strong> (list of <code class="xref py py-class docutils literal"><span class="pre">indra.statement.Agent</span></code>) &#8211; The list of molecular species being created by the conversion.</li>
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; Evidence objects in support of the synthesis statement.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Deacetylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Deacetylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Deacetylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Deacetylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Deacetylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.DecreaseAmount">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">DecreaseAmount</code><span class="sig-paren">(</span><em>subj</em>, <em>obj</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#DecreaseAmount"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.DecreaseAmount" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.RegulateAmount" title="indra.statements.RegulateAmount"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.RegulateAmount</span></code></a></p>
<p>Degradation of a protein, possibly mediated by another protein.</p>
<p>Note that this statement can also be used to represent inhibitors of
synthesis (e.g., cycloheximide).</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>subj</strong> (<em>:py:class`indra.statement.Agent`</em>) &#8211; The protein mediating the degradation.</li>
<li><strong>obj</strong> (<code class="xref py py-class docutils literal"><span class="pre">indra.statement.Agent</span></code>) &#8211; The protein that is degraded.</li>
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; Evidence objects in support of the degradation statement.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Defarnesylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Defarnesylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Defarnesylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Defarnesylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Defarnesylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Degeranylgeranylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Degeranylgeranylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Degeranylgeranylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Degeranylgeranylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Degeranylgeranylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Deglycosylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Deglycosylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Deglycosylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Deglycosylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Deglycosylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Dehydroxylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Dehydroxylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Dehydroxylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Dehydroxylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Dehydroxylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Demethylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Demethylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Demethylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Demethylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Demethylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Demyristoylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Demyristoylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Demyristoylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Demyristoylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Demyristoylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Depalmitoylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Depalmitoylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Depalmitoylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Depalmitoylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Depalmitoylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Dephosphorylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Dephosphorylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Dephosphorylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Dephosphorylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Dephosphorylation modification.</p>
<p class="rubric">Examples</p>
<p>DUSP6 dephosphorylates ERK (MAPK1) at T185:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">dusp6</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;DUSP6&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">erk</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAPK1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">dephos</span> <span class="o">=</span> <span class="n">Dephosphorylation</span><span class="p">(</span><span class="n">dusp6</span><span class="p">,</span> <span class="n">erk</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;185&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Deribosylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Deribosylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Deribosylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Deribosylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Deribosylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Desumoylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Desumoylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Desumoylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Desumoylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Desumoylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Deubiquitination">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Deubiquitination</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Deubiquitination"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Deubiquitination" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.RemoveModification</span></code></p>
<p>Deubiquitination modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Evidence">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Evidence</code><span class="sig-paren">(</span><em>source_api=None</em>, <em>source_id=None</em>, <em>pmid=None</em>, <em>text=None</em>, <em>annotations=None</em>, <em>epistemics=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Evidence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Evidence" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Container for evidence supporting a given statement.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>source_api</strong> (<em>str or None</em>) &#8211; String identifying the INDRA API used to capture the statement,
e.g., &#8216;trips&#8217;, &#8216;biopax&#8217;, &#8216;bel&#8217;.</li>
<li><strong>source_id</strong> (<em>str or None</em>) &#8211; For statements drawn from databases, ID of the database entity
corresponding to the statement.</li>
<li><strong>pmid</strong> (<em>str or None</em>) &#8211; String indicating the Pubmed ID of the source of the statement.</li>
<li><strong>text</strong> (<em>str</em>) &#8211; Natural language text supporting the statement.</li>
<li><strong>annotations</strong> (<em>dict</em>) &#8211; Dictionary containing additional information on the
context of the statement, e.g., species, cell line,
tissue type, etc. The entries may vary depending on
the source of the information.</li>
<li><strong>epistemics</strong> (<em>dict</em>) &#8211; A dictionary describing various forms of epistemic
certainty associated with the statement.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Farnesylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Farnesylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Farnesylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Farnesylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Farnesylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Gap">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Gap</code><span class="sig-paren">(</span><em>gap</em>, <em>ras</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Gap"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Gap" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Acceleration of a GTPase protein&#8217;s GTP hydrolysis rate by a GAP.</p>
<p>Represents the generic process by which a GTPase activating protein (GAP)
catalyzes GTP hydrolysis by a particular small GTPase protein.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gap</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The GTPase activating protein.</li>
<li><strong>ras</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The GTPase protein.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>RASA1 catalyzes GTP hydrolysis on KRAS:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">rasa1</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;RASA1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kras</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;KRAS&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gap</span> <span class="o">=</span> <span class="n">Gap</span><span class="p">(</span><span class="n">rasa1</span><span class="p">,</span> <span class="n">kras</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Gef">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Gef</code><span class="sig-paren">(</span><em>gef</em>, <em>ras</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Gef"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Gef" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Exchange of GTP for GDP on a small GTPase protein mediated by a GEF.</p>
<p>Represents the generic process by which a guanosine exchange factor (GEF)
catalyzes nucleotide exchange on a GTPase protein.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gef</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The guanosine exchange factor.</li>
<li><strong>ras</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The GTPase protein.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>SOS1 catalyzes nucleotide exchange on KRAS:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">sos</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;SOS1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">kras</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;KRAS&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">gef</span> <span class="o">=</span> <span class="n">Gef</span><span class="p">(</span><span class="n">sos</span><span class="p">,</span> <span class="n">kras</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Geranylgeranylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Geranylgeranylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Geranylgeranylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Geranylgeranylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Geranylgeranylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Glycosylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Glycosylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Glycosylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Glycosylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Glycosylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.HasActivity">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">HasActivity</code><span class="sig-paren">(</span><em>agent</em>, <em>activity</em>, <em>has_activity</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#HasActivity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.HasActivity" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>States that an Agent has or doesn&#8217;t have a given activity type.</p>
<p>With this Statement, one cane express that a given protein is a kinase, or,
for instance, that it is a transcription factor. It is also possible to
construct negative statements with which one epxresses, for instance,
that a given protein is not a kinase.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>agent</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The Agent that that statement is about. Note that the detailed state
of the Agent is not relevant for this type of statement.</li>
<li><strong>activity</strong> (<em>str</em>) &#8211; The type of activity, e.g., &#8220;kinase&#8221;.</li>
<li><strong>has_activity</strong> (<em>bool</em>) &#8211; Whether the given Agent has the given activity (True) or
not (False).</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Hydroxylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Hydroxylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Hydroxylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Hydroxylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Hydroxylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.IncreaseAmount">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">IncreaseAmount</code><span class="sig-paren">(</span><em>subj</em>, <em>obj</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#IncreaseAmount"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.IncreaseAmount" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.RegulateAmount" title="indra.statements.RegulateAmount"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.RegulateAmount</span></code></a></p>
<p>Synthesis of a protein, possibly mediated by another protein.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>subj</strong> (<em>:py:class`indra.statement.Agent`</em>) &#8211; The protein mediating the synthesis.</li>
<li><strong>obj</strong> (<code class="xref py py-class docutils literal"><span class="pre">indra.statement.Agent</span></code>) &#8211; The protein that is synthesized.</li>
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; Evidence objects in support of the synthesis statement.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Inhibition">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Inhibition</code><span class="sig-paren">(</span><em>subj</em>, <em>obj</em>, <em>obj_activity=u'activity'</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Inhibition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Inhibition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.RegulateActivity" title="indra.statements.RegulateActivity"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.RegulateActivity</span></code></a></p>
<p>Indicates that a protein inhibits or deactivates another protein.</p>
<p>This statement is intended to be used for physical interactions where the
mechanism of inhibition is not explicitly specified, which is often the
case for descriptions of mechanisms extracted from the literature.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>subj</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The agent responsible for the change in activity, i.e., the &#8220;upstream&#8221;
node.</li>
<li><strong>obj</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The agent whose activity is influenced by the subject, i.e., the
&#8220;downstream&#8221; node.</li>
<li><strong>obj_activity</strong> (<em>Optional[str]</em>) &#8211; The activity of the obj Agent that is affected, e.g., its &#8220;kinase&#8221;
activity.</li>
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; Evidence objects in support of the modification.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="exception">
<dt id="indra.statements.InvalidLocationError">
<em class="property">exception </em><code class="descclassname">indra.statements.</code><code class="descname">InvalidLocationError</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#InvalidLocationError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.InvalidLocationError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">exceptions.ValueError</span></code></p>
<p>Invalid cellular component name.</p>
</dd></dl>

<dl class="exception">
<dt id="indra.statements.InvalidResidueError">
<em class="property">exception </em><code class="descclassname">indra.statements.</code><code class="descname">InvalidResidueError</code><span class="sig-paren">(</span><em>name</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#InvalidResidueError"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.InvalidResidueError" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">exceptions.ValueError</span></code></p>
<p>Invalid residue (amino acid) name.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Methylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Methylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Methylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Methylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Methylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.ModCondition">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">ModCondition</code><span class="sig-paren">(</span><em>mod_type</em>, <em>residue=None</em>, <em>position=None</em>, <em>is_modified=True</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#ModCondition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.ModCondition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Post-translational modification state at an amino acid position.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>mod_type</strong> (<em>str</em>) &#8211; The type of post-translational modification, e.g., &#8216;phosphorylation&#8217;.
Valid modification types currently include: &#8216;phosphorylation&#8217;,
&#8216;ubiquitination&#8217;, &#8216;sumoylation&#8217;, &#8216;hydroxylation&#8217;, and &#8216;acetylation&#8217;.
If an invalid modification type is passed an InvalidModTypeError is
raised.</li>
<li><strong>residue</strong> (<em>str or None</em>) &#8211; String indicating the modified amino acid, e.g., &#8216;Y&#8217; or &#8216;tyrosine&#8217;.
If None, indicates that the residue at the modification site is
unknown or unspecified.</li>
<li><strong>position</strong> (<em>str or None</em>) &#8211; String indicating the position of the modified amino acid, e.g., &#8216;202&#8217;.
If None, indicates that the position is unknown or unspecified.</li>
<li><strong>is_modified</strong> (<em>bool</em>) &#8211; Specifies whether the modification is present or absent. Setting the
flag specifies that the Agent with the ModCondition is unmodified
at the site.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Doubly-phosphorylated MEK (MAP2K1):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">phospho_mek</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAP2K1&#39;</span><span class="p">,</span> <span class="n">mods</span><span class="o">=</span><span class="p">(</span>
<span class="gp">... </span><span class="n">ModCondition</span><span class="p">(</span><span class="s1">&#39;phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;202&#39;</span><span class="p">),</span>
<span class="gp">... </span><span class="n">ModCondition</span><span class="p">(</span><span class="s1">&#39;phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;S&#39;</span><span class="p">,</span> <span class="s1">&#39;204&#39;</span><span class="p">)))</span>
</pre></div>
</div>
<p>ERK (MAPK1) unphosphorylated at tyrosine 187:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">unphos_erk</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAPK1&#39;</span><span class="p">,</span> <span class="n">mods</span><span class="o">=</span><span class="p">(</span>
<span class="gp">... </span><span class="n">ModCondition</span><span class="p">(</span><span class="s1">&#39;phosphorylation&#39;</span><span class="p">,</span> <span class="s1">&#39;Y&#39;</span><span class="p">,</span> <span class="s1">&#39;187&#39;</span><span class="p">,</span> <span class="n">is_modified</span><span class="o">=</span><span class="kc">False</span><span class="p">)))</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Modification">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Modification</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Modification"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Modification" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Generic statement representing the modification of a protein.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>enz</strong> (<em>:py:class`indra.statement.Agent`</em>) &#8211; The enzyme involved in the modification.</li>
<li><strong>sub</strong> (<code class="xref py py-class docutils literal"><span class="pre">indra.statement.Agent</span></code>) &#8211; The substrate of the modification.</li>
<li><strong>residue</strong> (<em>str or None</em>) &#8211; The amino acid residue being modified, or None if it is unknown or
unspecified.</li>
<li><strong>position</strong> (<em>str or None</em>) &#8211; The position of the modified amino acid, or None if it is unknown or
unspecified.</li>
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; Evidence objects in support of the modification.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.MutCondition">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">MutCondition</code><span class="sig-paren">(</span><em>position</em>, <em>residue_from</em>, <em>residue_to=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#MutCondition"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.MutCondition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>Mutation state of an amino acid position of an Agent.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>position</strong> (<em>str</em>) &#8211; Residue position of the mutation in the protein sequence.</li>
<li><strong>residue_from</strong> (<em>str</em>) &#8211; Wild-type (unmodified) amino acid residue at the given position.</li>
<li><strong>residue_to</strong> (<em>str</em>) &#8211; Amino acid at the position resulting from the mutation.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<p class="rubric">Examples</p>
<p>Represent EGFR with a L858R mutation:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">egfr_mutant</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;EGFR&#39;</span><span class="p">,</span> <span class="n">mutations</span><span class="o">=</span><span class="p">(</span><span class="n">MutCondition</span><span class="p">(</span><span class="s1">&#39;858&#39;</span><span class="p">,</span> <span class="s1">&#39;L&#39;</span><span class="p">,</span> <span class="s1">&#39;R&#39;</span><span class="p">)))</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Myristoylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Myristoylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Myristoylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Myristoylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Myristoylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Palmitoylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Palmitoylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Palmitoylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Palmitoylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Palmitoylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Phosphorylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Phosphorylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Phosphorylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Phosphorylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Phosphorylation modification.</p>
<p class="rubric">Examples</p>
<p>MEK (MAP2K1) phosphorylates ERK (MAPK1) at threonine 185:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">mek</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAP2K1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">erk</span> <span class="o">=</span> <span class="n">Agent</span><span class="p">(</span><span class="s1">&#39;MAPK1&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">phos</span> <span class="o">=</span> <span class="n">Phosphorylation</span><span class="p">(</span><span class="n">mek</span><span class="p">,</span> <span class="n">erk</span><span class="p">,</span> <span class="s1">&#39;T&#39;</span><span class="p">,</span> <span class="s1">&#39;185&#39;</span><span class="p">)</span>
</pre></div>
</div>
</dd></dl>

<dl class="class">
<dt id="indra.statements.RegulateActivity">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">RegulateActivity</code><a class="reference internal" href="../_modules/indra/statements.html#RegulateActivity"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.RegulateActivity" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Regulation of activity.</p>
<p>This class implements shared functionality of Activation and Inhibition
statements and it should not be instantiated directly.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.RegulateAmount">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">RegulateAmount</code><span class="sig-paren">(</span><em>subj</em>, <em>obj</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#RegulateAmount"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.RegulateAmount" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Superclass handling operations on directed, two-element interactions.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Ribosylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Ribosylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Ribosylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Ribosylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Ribosylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.SelfModification">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">SelfModification</code><span class="sig-paren">(</span><em>enz</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#SelfModification"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.SelfModification" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>Generic statement representing the self-modification of a protein.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>enz</strong> (<em>:py:class`indra.statement.Agent`</em>) &#8211; The enzyme involved in the modification, which is also the substrate.</li>
<li><strong>residue</strong> (<em>str or None</em>) &#8211; The amino acid residue being modified, or None if it is unknown or
unspecified.</li>
<li><strong>position</strong> (<em>str or None</em>) &#8211; The position of the modified amino acid, or None if it is unknown or
unspecified.</li>
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; Evidence objects in support of the modification.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Statement">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Statement</code><span class="sig-paren">(</span><em>evidence=None</em>, <em>supports=None</em>, <em>supported_by=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Statement"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Statement" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">object</span></code></p>
<p>The parent class of all statements.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>evidence</strong> (list of <a class="reference internal" href="#indra.statements.Evidence" title="indra.statements.Evidence"><code class="xref py py-class docutils literal"><span class="pre">Evidence</span></code></a>) &#8211; If a list of Evidence objects is passed to the constructor, the
value is set to this list. If a bare Evidence object is passed,
it is enclosed in a list. If no evidence is passed (the default),
the value is set to an empty list.</li>
<li><strong>supports</strong> (list of <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">Statement</span></code></a>) &#8211; Statements that this Statement supports.</li>
<li><strong>supported_by</strong> (list of <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">Statement</span></code></a>) &#8211; Statements supported by this statement.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="method">
<dt id="indra.statements.Statement.to_graph">
<code class="descname">to_graph</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Statement.to_graph"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Statement.to_graph" title="Permalink to this definition">¶</a></dt>
<dd><p>Return Statement as a networkx graph.</p>
</dd></dl>

<dl class="method">
<dt id="indra.statements.Statement.to_json">
<code class="descname">to_json</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Statement.to_json"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Statement.to_json" title="Permalink to this definition">¶</a></dt>
<dd><p>Return serialized Statement as a json dict.</p>
</dd></dl>

</dd></dl>

<dl class="class">
<dt id="indra.statements.Sumoylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Sumoylation</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Sumoylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Sumoylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Sumoylation modification.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Translocation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Translocation</code><span class="sig-paren">(</span><em>agent</em>, <em>from_location=None</em>, <em>to_location=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Translocation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Translocation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
<p>The translocation of a molecular agent from one location to another.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>agent</strong> (<a class="reference internal" href="#indra.statements.Agent" title="indra.statements.Agent"><code class="xref py py-class docutils literal"><span class="pre">Agent</span></code></a>) &#8211; The agent which translocates.</li>
<li><strong>from_location</strong> (<em>Optional[str]</em>) &#8211; The location from which the agent translocates. This must
be a valid GO cellular component name (e.g. &#8220;cytoplasm&#8221;)
or ID (e.g. &#8220;<a class="reference external" href="GO:0005737">GO:0005737</a>&#8221;).</li>
<li><strong>to_location</strong> (<em>Optional[str]</em>) &#8211; The location to which the agent translocates. This must
be a valid GO cellular component name or ID.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Transphosphorylation">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Transphosphorylation</code><span class="sig-paren">(</span><em>enz</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Transphosphorylation"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Transphosphorylation" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#indra.statements.SelfModification" title="indra.statements.SelfModification"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.SelfModification</span></code></a></p>
<p>Autophosphorylation in <em>trans.</em></p>
<p>Transphosphorylation assumes that a kinase is already bound to a substrate
(usually of the same molecular species), and phosphorylates it in an
intra-molecular fashion. The enz property of the statement must have
exactly one bound_conditions entry, and we assume that enz phosphorylates
this molecule. The bound_neg property is ignored here.</p>
</dd></dl>

<dl class="class">
<dt id="indra.statements.Ubiquitination">
<em class="property">class </em><code class="descclassname">indra.statements.</code><code class="descname">Ubiquitination</code><span class="sig-paren">(</span><em>enz</em>, <em>sub</em>, <em>residue=None</em>, <em>position=None</em>, <em>evidence=None</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#Ubiquitination"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.Ubiquitination" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal"><span class="pre">indra.statements.AddModification</span></code></p>
<p>Ubiquitination modification.</p>
</dd></dl>

<dl class="function">
<dt id="indra.statements.get_valid_location">
<code class="descclassname">indra.statements.</code><code class="descname">get_valid_location</code><span class="sig-paren">(</span><em>location</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#get_valid_location"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.get_valid_location" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the given location represents a valid cellular component.</p>
</dd></dl>

<dl class="function">
<dt id="indra.statements.get_valid_residue">
<code class="descclassname">indra.statements.</code><code class="descname">get_valid_residue</code><span class="sig-paren">(</span><em>residue</em><span class="sig-paren">)</span><a class="reference internal" href="../_modules/indra/statements.html#get_valid_residue"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.statements.get_valid_residue" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if the given string represents a valid amino acid residue.</p>
</dd></dl>

<dl class="attribute">
<dt id="indra.statements.mt">
<code class="descclassname">indra.statements.</code><code class="descname">mt</code><a class="headerlink" href="#indra.statements.mt" title="Permalink to this definition">¶</a></dt>
<dd><p>alias of <a class="reference internal" href="#indra.statements.Methylation" title="indra.statements.Methylation"><code class="xref py py-class docutils literal"><span class="pre">Methylation</span></code></a></p>
</dd></dl>

</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="sources/index.html" class="btn btn-neutral float-right" title="Processors for model input (indra.sources)" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="INDRA modules reference" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, B. M. Gyori, J. A. Bachman.
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../',
            VERSION:'1.5.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../_static/jquery.js"></script>
      <script type="text/javascript" src="../_static/underscore.js"></script>
      <script type="text/javascript" src="../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>