

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>indra.databases.cbio_client &mdash; INDRA 1.5.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="INDRA 1.5.0 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="../../../index.html" class="icon icon-home"> INDRA
        

        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">License and funding</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#installing-python">Installing Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#installing-indra">Installing INDRA</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#installing-via-github">Installing via Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#cloning-the-source-code-from-github">Cloning the source code from Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#installing-releases-with-pip">Installing releases with pip</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#indra-dependencies">INDRA dependencies</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#pysb-and-bionetgen">PySB and BioNetGen</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#pyjnius">Pyjnius</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#graphviz">Graphviz</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#matplotlib">Matplotlib</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#optional-additional-dependencies">Optional additional dependencies</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting started with INDRA</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#importing-indra-and-its-modules">Importing INDRA and its modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#basic-usage-examples">Basic usage examples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#reading-a-sentence-with-trips">Reading a sentence with TRIPS</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#reading-a-pubmed-central-article-with-reach">Reading a PubMed Central article with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#getting-the-neighborhood-of-proteins-from-the-bel-large-corpus">Getting the neighborhood of proteins from the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#getting-paths-between-two-proteins-from-pathwaycommons-biopax">Getting paths between two proteins from PathwayCommons (BioPAX)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#constructing-indra-statements-manually">Constructing INDRA Statements manually</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#assembling-a-pysb-model-and-exporting-to-sbml">Assembling a PySB model and exporting to SBML</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">INDRA modules reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/statements.html">INDRA Statements (<code class="docutils literal"><span class="pre">indra.statements</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/sources/index.html">Processors for model input (<code class="docutils literal"><span class="pre">indra.sources</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/bel/index.html">BEL (<code class="docutils literal"><span class="pre">indra.sources.bel</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/bel/index.html#module-indra.sources.bel.bel_api">BEL API (<code class="docutils literal"><span class="pre">indra.sources.bel.bel_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/bel/index.html#module-indra.sources.bel.processor">BEL Processor (<code class="docutils literal"><span class="pre">indra.sources.bel.processor</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/biopax/index.html">Biopax (<code class="docutils literal"><span class="pre">indra.sources.biopax</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/biopax/index.html#module-indra.sources.biopax.biopax_api">Biopax API (<code class="docutils literal"><span class="pre">indra.sources.biopax.biopax_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/biopax/index.html#module-indra.sources.biopax.processor">Biopax Processor (<code class="docutils literal"><span class="pre">indra.sources.biopax.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/biopax/index.html#module-indra.sources.biopax.pathway_commons_client">Pathway Commons Client (<code class="docutils literal"><span class="pre">indra.sources.biopax.pathway_commons_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/reach/index.html">REACH (<code class="docutils literal"><span class="pre">indra.sources.reach</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/reach/index.html#module-indra.sources.reach.reach_api">REACH API (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/reach/index.html#module-indra.sources.reach.processor">REACH Processor (<code class="docutils literal"><span class="pre">indra.sources.reach.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/reach/index.html#module-indra.sources.reach.reach_reader">REACH reader (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_reader</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/trips/index.html">TRIPS (<code class="docutils literal"><span class="pre">indra.sources.trips</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/trips/index.html#module-indra.sources.trips.trips_api">TRIPS API (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/trips/index.html#module-indra.sources.trips.processor">TRIPS Processor (<code class="docutils literal"><span class="pre">indra.sources.trips.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/trips/index.html#module-indra.sources.trips.trips_client">TRIPS Client (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/ndex_cx/index.html">NDEx CX API (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx_api</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/ndex_cx/index.html#module-indra.sources.ndex_cx.processor">NDEx CX Processor (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx.processor</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/databases/index.html">Database clients (<code class="docutils literal"><span class="pre">indra.databases</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.hgnc_client">HGNC client (<code class="docutils literal"><span class="pre">indra.hgnc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.uniprot_client">Uniprot client (<code class="docutils literal"><span class="pre">indra.databases.uniprot_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.chebi_client">ChEBI client (<code class="docutils literal"><span class="pre">indra.databases.chebi_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.biogrid_client">BioGRID client (<code class="docutils literal"><span class="pre">indra.databases.biogrid_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.context_client">Cell type context client (<code class="docutils literal"><span class="pre">indra.databases.context_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.relevance_client">Network relevance client (<code class="docutils literal"><span class="pre">indra.databases.relevance_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.ndex_client">NDEx client (<code class="docutils literal"><span class="pre">indra.databases.ndex_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.cbio_client">cBio portal client (<code class="docutils literal"><span class="pre">indra.databases.cbio_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/literature/index.html">Literature clients (<code class="docutils literal"><span class="pre">indra.literature</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.pubmed_client">Pubmed client (<code class="docutils literal"><span class="pre">indra.literature.pubmed_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.pmc_client">Pubmed Central client (<code class="docutils literal"><span class="pre">indra.literature.pmc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.crossref_client">CrossRef client (<code class="docutils literal"><span class="pre">indra.literature.crossref_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.elsevier_client">Elsevier client (<code class="docutils literal"><span class="pre">indra.literature.elsevier_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/preassembler/index.html">Preassembly (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler">Preassembler (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler.grounding_mapper">Entity grounding curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.grounding_mapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler.sitemapper">Site curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.sitemapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler.hierarchy_manager">Hierarchy manager (<code class="docutils literal"><span class="pre">indra.preassembler.hierarchy_manager</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/belief/index.html">Belief Engine (<code class="docutils literal"><span class="pre">indra.belief</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/mechlinker/index.html">Mechanism Linker (<code class="docutils literal"><span class="pre">indra.mechlinker</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/assemblers/index.html">Assemblers of model output (<code class="docutils literal"><span class="pre">indra.assemblers</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/pysb_assembler.html">Executable PySB models (<code class="docutils literal"><span class="pre">indra.assemblers.pysb_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/cx_assembler.html">Cytoscape networks (<code class="docutils literal"><span class="pre">indra.assemblers.cx_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/english_assembler.html">Natural language (<code class="docutils literal"><span class="pre">indra.assemblers.english_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/graph_assembler.html">Node-edge graphs (<code class="docutils literal"><span class="pre">indra.assemblers.graph_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/sif_assembler.html">SIF / Boolean networks (<code class="docutils literal"><span class="pre">indra.assemblers.sif_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/index_card_assembler.html">MITRE &#8220;index cards&#8221; (<code class="docutils literal"><span class="pre">indra.assemblers.index_card_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/sbgn_assembler.html">SBGN output (<code class="docutils literal"><span class="pre">indra.assemblers.sbgn_assembler</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/explanation/index.html">Explanation (<code class="docutils literal"><span class="pre">indra.explanation</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/explanation/index.html#module-indra.explanation.model_checker">Check whether a rule-based model satisfies a property (<code class="docutils literal"><span class="pre">indra.explanation.model_checker</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/tools/index.html">Tools (<code class="docutils literal"><span class="pre">indra.tools</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.assemble_corpus">Run assembly components in a pipeline (<code class="docutils literal"><span class="pre">indra.tools.assemble_corpus</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.gene_network">Build a network from a gene list (<code class="docutils literal"><span class="pre">indra.tools.gene_network</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.executable_subnetwork">Build an executable model from a fragment of a large network (<code class="docutils literal"><span class="pre">indra.tools.executable_subnetwork</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.incremental_model">Build a model incrementally over time (<code class="docutils literal"><span class="pre">indra.tools.incremental_model</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/index.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/nl_modeling.html">Using natural language to build models</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/nl_modeling.html#read-indra-statements-from-a-natural-language-string">Read INDRA Statements from a natural language string</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/nl_modeling.html#assemble-the-indra-statements-into-a-rule-based-executable-model">Assemble the INDRA Statements into a rule-based executable model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/nl_modeling.html#exporting-the-model-into-other-common-formats">Exporting the model into other common formats</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/machine_reading.html">Large-Scale Machine Reading with Starcluster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#install-reach">Install REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#install-amazon-s3-support">Install Amazon S3 support</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#install-other-dependencies">Install other dependencies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#assemble-a-corpus-of-pmids">Assemble a Corpus of PMIDs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#process-the-papers-with-reach">Process the papers with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#extract-indra-statements-from-the-reach-output-on-s3">Extract INDRA Statements from the REACH output on S3</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#running-the-whole-pipeline-with-one-script">Running the whole pipeline with one script</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/machine_reading_aws.html">Large-Scale Machine Reading with Amazon Batch</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading_aws.html#how-it-works">How it Works</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/gene_network.html">Assembling everything known about a particular gene</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#collect-mechanisms-from-pathwaycommons-and-the-bel-large-corpus">Collect mechanisms from PathwayCommons and the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#collect-a-list-of-publications-that-discuss-the-gene-of-interest">Collect a list of publications that discuss the gene of interest</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#get-the-full-text-or-abstract-corresponding-to-the-publications">Get the full text or abstract corresponding to the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#read-the-content-of-the-publications">Read the content of the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#combine-all-statements-and-run-pre-assembly">Combine all statements and run pre-assembly</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#assemble-the-statements-into-a-network-model">Assemble the statements into a network model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../rest_api.html">REST API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../rest_api.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../rest_api.html#launching-the-rest-service">Launching the REST service</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../rest_api.html#documentation">Documentation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">INDRA</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      
    <li>indra.databases.cbio_client</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <h1>Source code for indra.databases.cbio_client</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">print_function</span><span class="p">,</span> <span class="n">unicode_literals</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="k">import</span> <span class="nb">dict</span><span class="p">,</span> <span class="nb">str</span>
<span class="kn">import</span> <span class="nn">pandas</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="k">import</span> <span class="n">defaultdict</span>
<span class="c1"># Python3</span>
<span class="k">try</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">functools</span> <span class="k">import</span> <span class="n">lru_cache</span>
    <span class="kn">from</span> <span class="nn">io</span> <span class="k">import</span> <span class="n">StringIO</span>
<span class="c1"># Python2</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="kn">from</span> <span class="nn">functools32</span> <span class="k">import</span> <span class="n">lru_cache</span>
    <span class="kn">from</span> <span class="nn">StringIO</span> <span class="k">import</span> <span class="n">StringIO</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;cbio&#39;</span><span class="p">)</span>

<span class="n">cbio_url</span> <span class="o">=</span> <span class="s1">&#39;http://www.cbioportal.org/webservice.do&#39;</span>
<span class="n">ccle_study</span> <span class="o">=</span> <span class="s1">&#39;cellline_ccle_broad&#39;</span>


<span class="nd">@lru_cache</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">kwargs</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a data frame from a web service request to cBio portal.</span>

<span class="sd">    Sends a web service requrest to the cBio portal with arguments given in</span>
<span class="sd">    the dictionary data and returns a Pandas data frame on success.</span>

<span class="sd">    More information about the service here:</span>
<span class="sd">    http://www.cbioportal.org/web_api.jsp</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    kwargs : dict</span>
<span class="sd">        A dict of parameters for the query. Entries map directly to web service</span>
<span class="sd">        calls with the exception of the optional &#39;skiprows&#39; entry, whose value</span>
<span class="sd">        is used as the number of rows to skip when reading the result data</span>
<span class="sd">        frame.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    df : pandas.DataFrame</span>
<span class="sd">        Response from cBioPortal as a Pandas DataFrame.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">skiprows</span> <span class="o">=</span> <span class="n">kwargs</span><span class="o">.</span><span class="n">pop</span><span class="p">(</span><span class="s1">&#39;skiprows&#39;</span><span class="p">,</span> <span class="kc">None</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">cbio_url</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">kwargs</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
        <span class="c1"># Adaptively skip rows based on number of comment lines</span>
        <span class="k">if</span> <span class="n">skiprows</span> <span class="o">==</span> <span class="o">-</span><span class="mi">1</span><span class="p">:</span>
            <span class="n">lines</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
            <span class="n">skiprows</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">for</span> <span class="n">line</span> <span class="ow">in</span> <span class="n">lines</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">line</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;#&#39;</span><span class="p">):</span>
                    <span class="n">skiprows</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">break</span>
        <span class="n">csv_StringIO</span> <span class="o">=</span> <span class="n">StringIO</span><span class="p">(</span><span class="n">res</span><span class="o">.</span><span class="n">text</span><span class="p">)</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">pandas</span><span class="o">.</span><span class="n">read_csv</span><span class="p">(</span><span class="n">csv_StringIO</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">,</span> <span class="n">skiprows</span><span class="o">=</span><span class="n">skiprows</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">df</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;Request returned with code </span><span class="si">%d</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">res</span><span class="o">.</span><span class="n">status_code</span><span class="p">)</span>


<div class="viewcode-block" id="get_mutations"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_mutations">[docs]</a><span class="k">def</span> <span class="nf">get_mutations</span><span class="p">(</span><span class="n">study_id</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">,</span> <span class="n">mutation_type</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span>
                  <span class="n">case_id</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return mutations as a list of genes and list of amino acid changes.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    study_id : str</span>
<span class="sd">        The ID of the cBio study.</span>
<span class="sd">        Example: &#39;cellline_ccle_broad&#39; or &#39;paad_icgc&#39;</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        A list of genes with their HGNC symbols.</span>
<span class="sd">        Example: [&#39;BRAF&#39;, &#39;KRAS&#39;]</span>
<span class="sd">    mutation_type : Optional[str]</span>
<span class="sd">        The type of mutation to filter to.</span>
<span class="sd">        mutation_type can be one of: missense, nonsense, frame_shift_ins,</span>
<span class="sd">        frame_shift_del, splice_site</span>
<span class="sd">    case_id : Optional[str]</span>
<span class="sd">        The case ID within the study to filter to.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mutations : tuple[list]</span>
<span class="sd">        A tuple of two lists, the first one containing a list of genes, and</span>
<span class="sd">        the second one a list of amino acid changes in those genes.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genetic_profile</span> <span class="o">=</span> <span class="n">get_genetic_profiles</span><span class="p">(</span><span class="n">study_id</span><span class="p">,</span> <span class="s1">&#39;mutation&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">gene_list_str</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span>

    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getMutationData&#39;</span><span class="p">,</span>
            <span class="s1">&#39;case_set_id&#39;</span><span class="p">:</span> <span class="n">study_id</span><span class="p">,</span>
            <span class="s1">&#39;genetic_profile_id&#39;</span><span class="p">:</span> <span class="n">genetic_profile</span><span class="p">,</span>
            <span class="s1">&#39;gene_list&#39;</span><span class="p">:</span> <span class="n">gene_list_str</span><span class="p">,</span>
            <span class="s1">&#39;skiprows&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">case_id</span><span class="p">:</span>
        <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;case_id&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">case_id</span><span class="p">]</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">_filter_data_frame</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;gene_symbol&#39;</span><span class="p">,</span> <span class="s1">&#39;amino_acid_change&#39;</span><span class="p">],</span>
                             <span class="s1">&#39;mutation_type&#39;</span><span class="p">,</span> <span class="n">mutation_type</span><span class="p">)</span>
    <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;gene_symbol&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">&#39;gene_symbol&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">()),</span>
                 <span class="s1">&#39;amino_acid_change&#39;</span><span class="p">:</span> <span class="nb">list</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">&#39;amino_acid_change&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())}</span>
    <span class="k">return</span> <span class="n">mutations</span></div>


<div class="viewcode-block" id="get_case_lists"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_case_lists">[docs]</a><span class="k">def</span> <span class="nf">get_case_lists</span><span class="p">(</span><span class="n">study_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of the case set ids for a particular study.</span>

<span class="sd">    TAKE NOTE the &quot;case_list_id&quot; are the same thing as &quot;case_set_id&quot;</span>
<span class="sd">    Within the data, this string is referred to as a &quot;case_list_id&quot;.</span>
<span class="sd">    Within API calls it is referred to as a &#39;case_set_id&#39;.</span>
<span class="sd">    The documentation does not make this explicitly clear.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    study_id : str</span>
<span class="sd">        The ID of the cBio study.</span>
<span class="sd">        Example: &#39;cellline_ccle_broad&#39; or &#39;paad_icgc&#39;</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    case_set_ids : dict[dict[int]]</span>
<span class="sd">        A dict keyed to cases containing a dict keyed to genes</span>
<span class="sd">        containing int</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getCaseLists&#39;</span><span class="p">,</span>
            <span class="s1">&#39;cancer_study_id&#39;</span><span class="p">:</span> <span class="n">study_id</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">case_set_ids</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;case_list_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">case_set_ids</span></div>


<div class="viewcode-block" id="get_profile_data"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_profile_data">[docs]</a><span class="k">def</span> <span class="nf">get_profile_data</span><span class="p">(</span><span class="n">study_id</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">,</span>
                     <span class="n">profile_filter</span><span class="p">,</span> <span class="n">case_set_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return dict of cases and genes and their respective values.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    study_id : str</span>
<span class="sd">        The ID of the cBio study.</span>
<span class="sd">        Example: &#39;cellline_ccle_broad&#39; or &#39;paad_icgc&#39;</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        A list of genes with their HGNC symbols.</span>
<span class="sd">        Example: [&#39;BRAF&#39;, &#39;KRAS&#39;]</span>
<span class="sd">    profile_filter : str</span>
<span class="sd">        A string used to filter the profiles to return. Will be one of:</span>
<span class="sd">        - MUTATION</span>
<span class="sd">        - MUTATION_EXTENDED</span>
<span class="sd">        - COPY_NUMBER_ALTERATION</span>
<span class="sd">        - MRNA_EXPRESSION</span>
<span class="sd">        - METHYLATION</span>
<span class="sd">    case_set_filter : Optional[str]</span>
<span class="sd">        A string that specifices which case_set_id to use, based on a complete</span>
<span class="sd">        or partial match. If not provided, will look for study_id + &#39;_all&#39;</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    profile_data : dict[dict[int]]</span>
<span class="sd">        A dict keyed to cases containing a dict keyed to genes</span>
<span class="sd">        containing int</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">genetic_profile</span> <span class="o">=</span> <span class="n">get_genetic_profiles</span><span class="p">(</span><span class="n">study_id</span><span class="p">,</span> <span class="n">profile_filter</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
    <span class="n">gene_list_str</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span>
    <span class="n">case_set_ids</span> <span class="o">=</span> <span class="n">get_case_lists</span><span class="p">(</span><span class="n">study_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">case_set_filter</span><span class="p">:</span>
        <span class="n">case_set_id</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">case_set_ids</span> <span class="k">if</span> <span class="n">case_set_filter</span> <span class="ow">in</span> <span class="n">x</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">case_set_id</span> <span class="o">=</span> <span class="n">study_id</span> <span class="o">+</span> <span class="s1">&#39;_all&#39;</span>
        <span class="c1"># based on looking at the cBioPortal, this is a common case_set_id</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getProfileData&#39;</span><span class="p">,</span>
            <span class="s1">&#39;case_set_id&#39;</span><span class="p">:</span> <span class="n">case_set_id</span><span class="p">,</span>
            <span class="s1">&#39;genetic_profile_id&#39;</span><span class="p">:</span> <span class="n">genetic_profile</span><span class="p">,</span>
            <span class="s1">&#39;gene_list&#39;</span><span class="p">:</span> <span class="n">gene_list_str</span><span class="p">,</span>
            <span class="s1">&#39;skiprows&#39;</span><span class="p">:</span> <span class="mi">2</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">case_list_df</span> <span class="o">=</span> <span class="p">[</span><span class="n">x</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
                    <span class="k">if</span> <span class="n">x</span> <span class="ow">not</span> <span class="ow">in</span> <span class="p">[</span><span class="s1">&#39;GENE_ID&#39;</span><span class="p">,</span> <span class="s1">&#39;COMMON&#39;</span><span class="p">]]</span>
    <span class="n">profile_data</span> <span class="o">=</span> <span class="p">{</span><span class="n">case</span><span class="p">:</span> <span class="p">{</span><span class="n">g</span><span class="p">:</span> <span class="kc">None</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">}</span>
                    <span class="k">for</span> <span class="n">case</span> <span class="ow">in</span> <span class="n">case_list_df</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">case</span> <span class="ow">in</span> <span class="n">case_list_df</span><span class="p">:</span>
        <span class="n">profile_values</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">case</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">g</span><span class="p">,</span> <span class="n">cv</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">profile_values</span><span class="p">):</span>
            <span class="k">if</span> <span class="ow">not</span> <span class="n">pandas</span><span class="o">.</span><span class="n">isnull</span><span class="p">(</span><span class="n">cv</span><span class="p">):</span>
                <span class="n">profile_data</span><span class="p">[</span><span class="n">case</span><span class="p">][</span><span class="n">g</span><span class="p">]</span> <span class="o">=</span> <span class="n">cv</span>
    <span class="k">return</span> <span class="n">profile_data</span></div>


<div class="viewcode-block" id="get_num_sequenced"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_num_sequenced">[docs]</a><span class="k">def</span> <span class="nf">get_num_sequenced</span><span class="p">(</span><span class="n">study_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return number of sequenced tumors for given study.</span>

<span class="sd">    This is useful for calculating mutation statistics in terms of the</span>
<span class="sd">    prevalence of certain mutations within a type of cancer.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    study_id : str</span>
<span class="sd">        The ID of the cBio study.</span>
<span class="sd">        Example: &#39;paad_icgc&#39;</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    num_case : int</span>
<span class="sd">        The number of sequenced tumors in the given study</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getCaseLists&#39;</span><span class="p">,</span>
            <span class="s1">&#39;cancer_study_id&#39;</span><span class="p">:</span> <span class="n">study_id</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">row_filter</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;case_list_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="s1">&#39;sequenced&#39;</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
    <span class="n">num_case</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">df</span><span class="p">[</span><span class="n">row_filter</span><span class="p">][</span><span class="s1">&#39;case_ids&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">tolist</span><span class="p">()[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39; &#39;</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">num_case</span></div>


<div class="viewcode-block" id="get_genetic_profiles"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_genetic_profiles">[docs]</a><span class="k">def</span> <span class="nf">get_genetic_profiles</span><span class="p">(</span><span class="n">study_id</span><span class="p">,</span> <span class="n">profile_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return all the genetic profiles (data sets) for a given study.</span>

<span class="sd">    Genetic profiles are different types of data for a given study. For</span>
<span class="sd">    instance the study &#39;cellline_ccle_broad&#39; has profiles such as</span>
<span class="sd">    &#39;cellline_ccle_broad_mutations&#39; for mutations, &#39;cellline_ccle_broad_CNA&#39;</span>
<span class="sd">    for copy number alterations, etc.</span>


<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    study_id : str</span>
<span class="sd">        The ID of the cBio study.</span>
<span class="sd">        Example: &#39;paad_icgc&#39;</span>
<span class="sd">    profile_filter : Optional[str]</span>
<span class="sd">        A string used to filter the profiles to return.</span>
<span class="sd">        Will be one of:</span>
<span class="sd">        - MUTATION</span>
<span class="sd">        - MUTATION_EXTENDED</span>
<span class="sd">        - COPY_NUMBER_ALTERATION</span>
<span class="sd">        - MRNA_EXPRESSION</span>
<span class="sd">        - METHYLATION</span>
<span class="sd">        The genetic profiles can include &quot;mutation&quot;, &quot;CNA&quot;, &quot;rppa&quot;,</span>
<span class="sd">        &quot;methylation&quot;, etc.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    genetic_profiles : list[str]</span>
<span class="sd">        A list of genetic profiles available  for the given study.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getGeneticProfiles&#39;</span><span class="p">,</span>
            <span class="s1">&#39;cancer_study_id&#39;</span><span class="p">:</span> <span class="n">study_id</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">_filter_data_frame</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;genetic_profile_id&#39;</span><span class="p">],</span>
                             <span class="s1">&#39;genetic_alteration_type&#39;</span><span class="p">,</span> <span class="n">profile_filter</span><span class="p">)</span>
    <span class="n">genetic_profiles</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">&#39;genetic_profile_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">genetic_profiles</span></div>


<div class="viewcode-block" id="get_cancer_studies"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_cancer_studies">[docs]</a><span class="k">def</span> <span class="nf">get_cancer_studies</span><span class="p">(</span><span class="n">study_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of cancer study identifiers, optionally filtered.</span>

<span class="sd">    There are typically multiple studies for a given type of cancer and</span>
<span class="sd">    a filter can be used to constrain the returned list.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    study_filter : Optional[str]</span>
<span class="sd">        A string used to filter the study IDs to return. Example: &quot;paad&quot;</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    study_ids : list[str]</span>
<span class="sd">        A list of study IDs.</span>
<span class="sd">        For instance &quot;paad&quot; as a filter would result in a list</span>
<span class="sd">        of study IDs with paad in their name like &quot;paad_icgc&quot;, &quot;paad_tcga&quot;,</span>
<span class="sd">        etc.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getCancerStudies&#39;</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">_filter_data_frame</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;cancer_study_id&#39;</span><span class="p">],</span>
                             <span class="s1">&#39;cancer_study_id&#39;</span><span class="p">,</span> <span class="n">study_filter</span><span class="p">)</span>
    <span class="n">study_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">&#39;cancer_study_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">study_ids</span></div>


<div class="viewcode-block" id="get_cancer_types"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_cancer_types">[docs]</a><span class="k">def</span> <span class="nf">get_cancer_types</span><span class="p">(</span><span class="n">cancer_filter</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a list of cancer types, optionally filtered.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    cancer_filter : Optional[str]</span>
<span class="sd">        A string used to filter cancer types. Its value is the name or</span>
<span class="sd">        part of the name of a type of cancer. Example: &quot;melanoma&quot;,</span>
<span class="sd">        &quot;pancreatic&quot;, &quot;non-small cell lung&quot;</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    type_ids : list[str]</span>
<span class="sd">        A list of cancer types matching the filter.</span>
<span class="sd">        Example: for cancer_filter=&quot;pancreatic&quot;, the result includes</span>
<span class="sd">        &quot;panet&quot; (neuro-endocrine) and &quot;paad&quot; (adenocarcinoma)</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getTypesOfCancer&#39;</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">_filter_data_frame</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="p">[</span><span class="s1">&#39;type_of_cancer_id&#39;</span><span class="p">],</span> <span class="s1">&#39;name&#39;</span><span class="p">,</span> <span class="n">cancer_filter</span><span class="p">)</span>
    <span class="n">type_ids</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">res</span><span class="p">[</span><span class="s1">&#39;type_of_cancer_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">values</span><span class="p">())</span>
    <span class="k">return</span> <span class="n">type_ids</span></div>


<div class="viewcode-block" id="get_mutations_ccle"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_mutations_ccle">[docs]</a><span class="k">def</span> <span class="nf">get_mutations_ccle</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">cell_lines</span><span class="p">,</span> <span class="n">mutation_type</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a dict of mutations in given genes and cell lines from CCLE.</span>

<span class="sd">    This is a specialized call to get_mutations tailored to CCLE cell lines.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        A list of HGNC gene symbols to get mutations in</span>
<span class="sd">    cell_lines : list[str]</span>
<span class="sd">        A list of CCLE cell line names to get mutations for.</span>
<span class="sd">    mutation_type : Optional[str]</span>
<span class="sd">        The type of mutation to filter to.</span>
<span class="sd">        mutation_type can be one of: missense, nonsense, frame_shift_ins,</span>
<span class="sd">        frame_shift_del, splice_site</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mutations : dict</span>
<span class="sd">        The result from cBioPortal as a dict in the format</span>
<span class="sd">        {cell_line : {gene : [mutation1, mutation2, ...] }}</span>

<span class="sd">        Example:</span>
<span class="sd">        {&#39;LOXIMVI_SKIN&#39;: {&#39;BRAF&#39;: [&#39;V600E&#39;, &#39;I208V&#39;]},</span>
<span class="sd">        &#39;SKMEL30_SKIN&#39;: {&#39;BRAF&#39;: [&#39;D287H&#39;, &#39;E275K&#39;]}}</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mutations</span> <span class="o">=</span> <span class="p">{</span><span class="n">cl</span><span class="p">:</span> <span class="p">{</span><span class="n">g</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">}</span> <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="n">cell_lines</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">cell_line</span> <span class="ow">in</span> <span class="n">cell_lines</span><span class="p">:</span>
        <span class="n">mutations_cl</span> <span class="o">=</span> <span class="n">get_mutations</span><span class="p">(</span><span class="n">ccle_study</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">,</span>
                                     <span class="n">mutation_type</span><span class="o">=</span><span class="n">mutation_type</span><span class="p">,</span>
                                     <span class="n">case_id</span><span class="o">=</span><span class="n">cell_line</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">gene</span><span class="p">,</span> <span class="n">aa_change</span> <span class="ow">in</span> <span class="nb">zip</span><span class="p">(</span><span class="n">mutations_cl</span><span class="p">[</span><span class="s1">&#39;gene_symbol&#39;</span><span class="p">],</span>
                                   <span class="n">mutations_cl</span><span class="p">[</span><span class="s1">&#39;amino_acid_change&#39;</span><span class="p">]):</span>
            <span class="n">aa_change</span> <span class="o">=</span> <span class="nb">str</span><span class="p">(</span><span class="n">aa_change</span><span class="p">)</span>
            <span class="n">mutations</span><span class="p">[</span><span class="n">cell_line</span><span class="p">][</span><span class="n">gene</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">aa_change</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">mutations</span></div>


<div class="viewcode-block" id="get_ccle_lines_for_mutation"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_ccle_lines_for_mutation">[docs]</a><span class="k">def</span> <span class="nf">get_ccle_lines_for_mutation</span><span class="p">(</span><span class="n">gene</span><span class="p">,</span> <span class="n">amino_acid_change</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return cell lines with a given point mutation in a given gene.</span>

<span class="sd">    Checks which cell lines in CCLE have a particular point mutation</span>
<span class="sd">    in a given gene and return their names in a list.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene : str</span>
<span class="sd">        The HGNC symbol of the mutated gene in whose product the amino</span>
<span class="sd">        acid change occurs. Example: &quot;BRAF&quot;</span>
<span class="sd">    amino_acid_change : str</span>
<span class="sd">        The amino acid change of interest. Example: &quot;V600E&quot;</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    cell_lines : list</span>
<span class="sd">        A list of CCLE cell lines in which the given mutation occurs.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getMutationData&#39;</span><span class="p">,</span>
            <span class="s1">&#39;case_set_id&#39;</span><span class="p">:</span> <span class="n">ccle_study</span><span class="p">,</span>
            <span class="s1">&#39;genetic_profile_id&#39;</span><span class="p">:</span> <span class="n">ccle_study</span> <span class="o">+</span> <span class="s1">&#39;_mutations&#39;</span><span class="p">,</span>
            <span class="s1">&#39;gene_list&#39;</span><span class="p">:</span> <span class="n">gene</span><span class="p">,</span>
            <span class="s1">&#39;skiprows&#39;</span><span class="p">:</span> <span class="mi">1</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;amino_acid_change&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">amino_acid_change</span><span class="p">]</span>
    <span class="n">cell_lines</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="s1">&#39;case_id&#39;</span><span class="p">]</span><span class="o">.</span><span class="n">unique</span><span class="p">()</span><span class="o">.</span><span class="n">tolist</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">cell_lines</span></div>


<div class="viewcode-block" id="get_ccle_cna"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_ccle_cna">[docs]</a><span class="k">def</span> <span class="nf">get_ccle_cna</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">cell_lines</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a dict of CNAs in given genes and cell lines from CCLE.</span>

<span class="sd">    This is a specialized call to get_mutations tailored to CCLE cell lines.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        A list of HGNC gene symbols to get mutations in</span>
<span class="sd">    cell_lines : list[str]</span>
<span class="sd">        A list of CCLE cell line names to get mutations for.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    profile_data : dict[dict[int]]</span>
<span class="sd">        A dict keyed to cases containing a dict keyed to genes</span>
<span class="sd">        containing int</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">profile_data</span> <span class="o">=</span> <span class="n">get_profile_data</span><span class="p">(</span><span class="n">ccle_study</span><span class="p">,</span> <span class="n">gene_list</span><span class="p">,</span>
                                    <span class="s1">&#39;COPY_NUMBER_ALTERATION&#39;</span><span class="p">,</span> <span class="s1">&#39;all&#39;</span><span class="p">)</span>
    <span class="n">profile_data</span> <span class="o">=</span> <span class="nb">dict</span><span class="p">((</span><span class="n">key</span><span class="p">,</span> <span class="n">value</span><span class="p">)</span> <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">profile_data</span><span class="o">.</span><span class="n">items</span><span class="p">()</span>
                        <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">cell_lines</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">profile_data</span></div>


<div class="viewcode-block" id="get_ccle_mrna"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.cbio_client.get_ccle_mrna">[docs]</a><span class="k">def</span> <span class="nf">get_ccle_mrna</span><span class="p">(</span><span class="n">gene_list</span><span class="p">,</span> <span class="n">cell_lines</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a dict of mRNA amounts in given genes and cell lines from CCLE.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    gene_list : list[str]</span>
<span class="sd">        A list of HGNC gene symbols to get mRNA amounts for.</span>
<span class="sd">    cell_lines : list[str]</span>
<span class="sd">        A list of CCLE cell line names to get mRNA amounts for.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mrna_amounts : dict[dict[float]]</span>
<span class="sd">        A dict keyed to cell lines containing a dict keyed to genes</span>
<span class="sd">        containing float</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">gene_list_str</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;cmd&#39;</span><span class="p">:</span> <span class="s1">&#39;getProfileData&#39;</span><span class="p">,</span>
            <span class="s1">&#39;case_set_id&#39;</span><span class="p">:</span> <span class="n">ccle_study</span> <span class="o">+</span> <span class="s1">&#39;_mrna&#39;</span><span class="p">,</span>
            <span class="s1">&#39;genetic_profile_id&#39;</span><span class="p">:</span> <span class="n">ccle_study</span> <span class="o">+</span> <span class="s1">&#39;_mrna&#39;</span><span class="p">,</span>
            <span class="s1">&#39;gene_list&#39;</span><span class="p">:</span> <span class="n">gene_list_str</span><span class="p">,</span>
            <span class="s1">&#39;skiprows&#39;</span><span class="p">:</span> <span class="o">-</span><span class="mi">1</span><span class="p">}</span>
    <span class="n">df</span> <span class="o">=</span> <span class="n">send_request</span><span class="p">(</span><span class="o">**</span><span class="n">data</span><span class="p">)</span>
    <span class="n">mrna_amounts</span> <span class="o">=</span> <span class="p">{</span><span class="n">cl</span><span class="p">:</span> <span class="p">{</span><span class="n">g</span><span class="p">:</span> <span class="p">[]</span> <span class="k">for</span> <span class="n">g</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">}</span> <span class="k">for</span> <span class="n">cl</span> <span class="ow">in</span> <span class="n">cell_lines</span><span class="p">}</span>
    <span class="k">for</span> <span class="n">cell_line</span> <span class="ow">in</span> <span class="n">cell_lines</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">cell_line</span> <span class="ow">in</span> <span class="n">df</span><span class="o">.</span><span class="n">columns</span><span class="p">:</span>
            <span class="k">for</span> <span class="n">gene</span> <span class="ow">in</span> <span class="n">gene_list</span><span class="p">:</span>
                <span class="n">value_cell</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">cell_line</span><span class="p">][</span><span class="n">df</span><span class="p">[</span><span class="s1">&#39;COMMON&#39;</span><span class="p">]</span> <span class="o">==</span> <span class="n">gene</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">value_cell</span><span class="o">.</span><span class="n">empty</span><span class="p">:</span>
                    <span class="n">mrna_amounts</span><span class="p">[</span><span class="n">cell_line</span><span class="p">][</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">value</span> <span class="o">=</span> <span class="n">value_cell</span><span class="o">.</span><span class="n">values</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">mrna_amounts</span><span class="p">[</span><span class="n">cell_line</span><span class="p">][</span><span class="n">gene</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">mrna_amounts</span><span class="p">[</span><span class="n">cell_line</span><span class="p">]</span> <span class="o">=</span> <span class="kc">None</span>
    <span class="k">return</span> <span class="n">mrna_amounts</span></div>


<span class="k">def</span> <span class="nf">_filter_data_frame</span><span class="p">(</span><span class="n">df</span><span class="p">,</span> <span class="n">data_col</span><span class="p">,</span> <span class="n">filter_col</span><span class="p">,</span> <span class="n">filter_str</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a filtered data frame as a dictionary.&quot;&quot;&quot;</span>
    <span class="k">if</span> <span class="n">filter_str</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">relevant_cols</span> <span class="o">=</span> <span class="n">data_col</span> <span class="o">+</span> <span class="p">[</span><span class="n">filter_col</span><span class="p">]</span>
        <span class="n">df</span><span class="o">.</span><span class="n">dropna</span><span class="p">(</span><span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">subset</span><span class="o">=</span><span class="n">relevant_cols</span><span class="p">)</span>
        <span class="n">row_filter</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">filter_col</span><span class="p">]</span><span class="o">.</span><span class="n">str</span><span class="o">.</span><span class="n">contains</span><span class="p">(</span><span class="n">filter_str</span><span class="p">,</span> <span class="n">case</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
        <span class="n">data_list</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">row_filter</span><span class="p">][</span><span class="n">data_col</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">data_list</span> <span class="o">=</span> <span class="n">df</span><span class="p">[</span><span class="n">data_col</span><span class="p">]</span><span class="o">.</span><span class="n">to_dict</span><span class="p">()</span>
    <span class="k">return</span> <span class="n">data_list</span>
</pre></div>

          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, B. M. Gyori, J. A. Bachman.
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'1.5.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>