

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>indra.databases.uniprot_client &mdash; INDRA 1.5.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="INDRA 1.5.0 documentation" href="../../../index.html"/>
        <link rel="up" title="Module code" href="../../index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="../../../index.html" class="icon icon-home"> INDRA
        

        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul>
<li class="toctree-l1"><a class="reference internal" href="../../../license.html">License and funding</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#installing-python">Installing Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#installing-indra">Installing INDRA</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#installing-via-github">Installing via Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#cloning-the-source-code-from-github">Cloning the source code from Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#installing-releases-with-pip">Installing releases with pip</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../installation.html#indra-dependencies">INDRA dependencies</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#pysb-and-bionetgen">PySB and BioNetGen</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#pyjnius">Pyjnius</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#graphviz">Graphviz</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#matplotlib">Matplotlib</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../installation.html#optional-additional-dependencies">Optional additional dependencies</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting started with INDRA</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#importing-indra-and-its-modules">Importing INDRA and its modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../getting_started.html#basic-usage-examples">Basic usage examples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#reading-a-sentence-with-trips">Reading a sentence with TRIPS</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#reading-a-pubmed-central-article-with-reach">Reading a PubMed Central article with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#getting-the-neighborhood-of-proteins-from-the-bel-large-corpus">Getting the neighborhood of proteins from the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#getting-paths-between-two-proteins-from-pathwaycommons-biopax">Getting paths between two proteins from PathwayCommons (BioPAX)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#constructing-indra-statements-manually">Constructing INDRA Statements manually</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../getting_started.html#assembling-a-pysb-model-and-exporting-to-sbml">Assembling a PySB model and exporting to SBML</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules/index.html">INDRA modules reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/statements.html">INDRA Statements (<code class="docutils literal"><span class="pre">indra.statements</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/sources/index.html">Processors for model input (<code class="docutils literal"><span class="pre">indra.sources</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/bel/index.html">BEL (<code class="docutils literal"><span class="pre">indra.sources.bel</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/bel/index.html#module-indra.sources.bel.bel_api">BEL API (<code class="docutils literal"><span class="pre">indra.sources.bel.bel_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/bel/index.html#module-indra.sources.bel.processor">BEL Processor (<code class="docutils literal"><span class="pre">indra.sources.bel.processor</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/biopax/index.html">Biopax (<code class="docutils literal"><span class="pre">indra.sources.biopax</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/biopax/index.html#module-indra.sources.biopax.biopax_api">Biopax API (<code class="docutils literal"><span class="pre">indra.sources.biopax.biopax_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/biopax/index.html#module-indra.sources.biopax.processor">Biopax Processor (<code class="docutils literal"><span class="pre">indra.sources.biopax.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/biopax/index.html#module-indra.sources.biopax.pathway_commons_client">Pathway Commons Client (<code class="docutils literal"><span class="pre">indra.sources.biopax.pathway_commons_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/reach/index.html">REACH (<code class="docutils literal"><span class="pre">indra.sources.reach</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/reach/index.html#module-indra.sources.reach.reach_api">REACH API (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/reach/index.html#module-indra.sources.reach.processor">REACH Processor (<code class="docutils literal"><span class="pre">indra.sources.reach.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/reach/index.html#module-indra.sources.reach.reach_reader">REACH reader (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_reader</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/trips/index.html">TRIPS (<code class="docutils literal"><span class="pre">indra.sources.trips</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/trips/index.html#module-indra.sources.trips.trips_api">TRIPS API (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/trips/index.html#module-indra.sources.trips.processor">TRIPS Processor (<code class="docutils literal"><span class="pre">indra.sources.trips.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/trips/index.html#module-indra.sources.trips.trips_client">TRIPS Client (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/sources/ndex_cx/index.html">NDEx CX API (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx_api</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../../../modules/sources/ndex_cx/index.html#module-indra.sources.ndex_cx.processor">NDEx CX Processor (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx.processor</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/databases/index.html">Database clients (<code class="docutils literal"><span class="pre">indra.databases</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.hgnc_client">HGNC client (<code class="docutils literal"><span class="pre">indra.hgnc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.uniprot_client">Uniprot client (<code class="docutils literal"><span class="pre">indra.databases.uniprot_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.chebi_client">ChEBI client (<code class="docutils literal"><span class="pre">indra.databases.chebi_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.biogrid_client">BioGRID client (<code class="docutils literal"><span class="pre">indra.databases.biogrid_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.context_client">Cell type context client (<code class="docutils literal"><span class="pre">indra.databases.context_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.relevance_client">Network relevance client (<code class="docutils literal"><span class="pre">indra.databases.relevance_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.ndex_client">NDEx client (<code class="docutils literal"><span class="pre">indra.databases.ndex_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/databases/index.html#module-indra.databases.cbio_client">cBio portal client (<code class="docutils literal"><span class="pre">indra.databases.cbio_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/literature/index.html">Literature clients (<code class="docutils literal"><span class="pre">indra.literature</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.pubmed_client">Pubmed client (<code class="docutils literal"><span class="pre">indra.literature.pubmed_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.pmc_client">Pubmed Central client (<code class="docutils literal"><span class="pre">indra.literature.pmc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.crossref_client">CrossRef client (<code class="docutils literal"><span class="pre">indra.literature.crossref_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/literature/index.html#module-indra.literature.elsevier_client">Elsevier client (<code class="docutils literal"><span class="pre">indra.literature.elsevier_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/preassembler/index.html">Preassembly (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler">Preassembler (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler.grounding_mapper">Entity grounding curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.grounding_mapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler.sitemapper">Site curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.sitemapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/preassembler/index.html#module-indra.preassembler.hierarchy_manager">Hierarchy manager (<code class="docutils literal"><span class="pre">indra.preassembler.hierarchy_manager</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/belief/index.html">Belief Engine (<code class="docutils literal"><span class="pre">indra.belief</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/mechlinker/index.html">Mechanism Linker (<code class="docutils literal"><span class="pre">indra.mechlinker</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/assemblers/index.html">Assemblers of model output (<code class="docutils literal"><span class="pre">indra.assemblers</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/pysb_assembler.html">Executable PySB models (<code class="docutils literal"><span class="pre">indra.assemblers.pysb_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/cx_assembler.html">Cytoscape networks (<code class="docutils literal"><span class="pre">indra.assemblers.cx_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/english_assembler.html">Natural language (<code class="docutils literal"><span class="pre">indra.assemblers.english_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/graph_assembler.html">Node-edge graphs (<code class="docutils literal"><span class="pre">indra.assemblers.graph_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/sif_assembler.html">SIF / Boolean networks (<code class="docutils literal"><span class="pre">indra.assemblers.sif_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/index_card_assembler.html">MITRE &#8220;index cards&#8221; (<code class="docutils literal"><span class="pre">indra.assemblers.index_card_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/assemblers/sbgn_assembler.html">SBGN output (<code class="docutils literal"><span class="pre">indra.assemblers.sbgn_assembler</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/explanation/index.html">Explanation (<code class="docutils literal"><span class="pre">indra.explanation</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/explanation/index.html#module-indra.explanation.model_checker">Check whether a rule-based model satisfies a property (<code class="docutils literal"><span class="pre">indra.explanation.model_checker</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../modules/tools/index.html">Tools (<code class="docutils literal"><span class="pre">indra.tools</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.assemble_corpus">Run assembly components in a pipeline (<code class="docutils literal"><span class="pre">indra.tools.assemble_corpus</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.gene_network">Build a network from a gene list (<code class="docutils literal"><span class="pre">indra.tools.gene_network</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.executable_subnetwork">Build an executable model from a fragment of a large network (<code class="docutils literal"><span class="pre">indra.tools.executable_subnetwork</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../modules/tools/index.html#module-indra.tools.incremental_model">Build a model incrementally over time (<code class="docutils literal"><span class="pre">indra.tools.incremental_model</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../tutorials/index.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/nl_modeling.html">Using natural language to build models</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/nl_modeling.html#read-indra-statements-from-a-natural-language-string">Read INDRA Statements from a natural language string</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/nl_modeling.html#assemble-the-indra-statements-into-a-rule-based-executable-model">Assemble the INDRA Statements into a rule-based executable model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/nl_modeling.html#exporting-the-model-into-other-common-formats">Exporting the model into other common formats</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/machine_reading.html">Large-Scale Machine Reading with Starcluster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#install-reach">Install REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#install-amazon-s3-support">Install Amazon S3 support</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#install-other-dependencies">Install other dependencies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#assemble-a-corpus-of-pmids">Assemble a Corpus of PMIDs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#process-the-papers-with-reach">Process the papers with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#extract-indra-statements-from-the-reach-output-on-s3">Extract INDRA Statements from the REACH output on S3</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading.html#running-the-whole-pipeline-with-one-script">Running the whole pipeline with one script</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/machine_reading_aws.html">Large-Scale Machine Reading with Amazon Batch</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/machine_reading_aws.html#how-it-works">How it Works</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../../tutorials/gene_network.html">Assembling everything known about a particular gene</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#collect-mechanisms-from-pathwaycommons-and-the-bel-large-corpus">Collect mechanisms from PathwayCommons and the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#collect-a-list-of-publications-that-discuss-the-gene-of-interest">Collect a list of publications that discuss the gene of interest</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#get-the-full-text-or-abstract-corresponding-to-the-publications">Get the full text or abstract corresponding to the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#read-the-content-of-the-publications">Read the content of the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#combine-all-statements-and-run-pre-assembly">Combine all statements and run pre-assembly</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../../tutorials/gene_network.html#assemble-the-statements-into-a-network-model">Assemble the statements into a network model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../../rest_api.html">REST API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../../rest_api.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../rest_api.html#launching-the-rest-service">Launching the REST service</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../../rest_api.html#documentation">Documentation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../../index.html">INDRA</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../../index.html">Module code</a> &raquo;</li>
      
    <li>indra.databases.uniprot_client</li>
      <li class="wy-breadcrumbs-aside">
        
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <h1>Source code for indra.databases.uniprot_client</h1><div class="highlight"><pre>
<span></span><span class="kn">from</span> <span class="nn">__future__</span> <span class="k">import</span> <span class="n">absolute_import</span><span class="p">,</span> <span class="n">print_function</span><span class="p">,</span> <span class="n">unicode_literals</span>
<span class="kn">from</span> <span class="nn">builtins</span> <span class="k">import</span> <span class="nb">dict</span><span class="p">,</span> <span class="nb">str</span>
<span class="kn">import</span> <span class="nn">os</span>
<span class="kn">import</span> <span class="nn">csv</span>
<span class="kn">import</span> <span class="nn">rdflib</span>
<span class="kn">import</span> <span class="nn">logging</span>
<span class="kn">import</span> <span class="nn">requests</span>
<span class="k">try</span><span class="p">:</span>
    <span class="c1"># Python 3</span>
    <span class="kn">from</span> <span class="nn">functools</span> <span class="k">import</span> <span class="n">lru_cache</span>
    <span class="kn">from</span> <span class="nn">urllib.error</span> <span class="k">import</span> <span class="n">HTTPError</span>
<span class="k">except</span> <span class="ne">ImportError</span><span class="p">:</span>
    <span class="c1"># Python 2</span>
    <span class="kn">from</span> <span class="nn">functools32</span> <span class="k">import</span> <span class="n">lru_cache</span>
    <span class="kn">from</span> <span class="nn">urllib2</span> <span class="k">import</span> <span class="n">HTTPError</span>
<span class="kn">from</span> <span class="nn">indra.util</span> <span class="k">import</span> <span class="n">read_unicode_csv</span>

<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">(</span><span class="s1">&#39;uniprot&#39;</span><span class="p">)</span>

<span class="n">uniprot_url</span> <span class="o">=</span> <span class="s1">&#39;http://www.uniprot.org/uniprot/&#39;</span>

<span class="n">rdf_prefixes</span> <span class="o">=</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">    PREFIX up: &lt;http://purl.uniprot.org/core/&gt;</span>
<span class="s2">    PREFIX db: &lt;http://purl.uniprot.org/database/&gt;</span>
<span class="s2">    PREFIX faldo: &lt;http://biohackathon.org/resource/faldo#&gt;</span>
<span class="s2">    PREFIX rdf: &lt;http://www.w3.org/1999/02/22-rdf-syntax-ns#&gt;</span>
<span class="s2">    PREFIX rdfs: &lt;http://www.w3.org/2000/01/rdf-schema#&gt; &quot;&quot;&quot;</span>


<span class="nd">@lru_cache</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="mi">10000</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt entry as an RDF graph for the given UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID to be queried.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    g : rdflib.Graph</span>
<span class="sd">        The RDF graph corresponding to the UniProt entry.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="c1"># Try looking up a primary ID if the given one</span>
    <span class="c1"># is a secondary ID</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">prim_ids</span> <span class="o">=</span> <span class="n">uniprot_sec</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="n">protein_id</span> <span class="o">=</span> <span class="n">prim_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="n">url</span> <span class="o">=</span> <span class="n">uniprot_url</span> <span class="o">+</span> <span class="n">protein_id</span> <span class="o">+</span> <span class="s1">&#39;.rdf&#39;</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">rdflib</span><span class="o">.</span><span class="n">Graph</span><span class="p">()</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">g</span><span class="o">.</span><span class="n">parse</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="k">except</span> <span class="n">HTTPError</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Could not find protein with id </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">protein_id</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="k">except</span> <span class="n">rdflib</span><span class="o">.</span><span class="n">exceptions</span><span class="o">.</span><span class="n">ParserError</span> <span class="k">as</span> <span class="n">e</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s1">&#39;Could not parse RDF at </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">url</span><span class="p">)</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="n">e</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="c1"># Check if the entry has been replaced by a new entry</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?res2</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?res1 up:replacedBy ?res2 .</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">term</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">replaced_by_id</span> <span class="o">=</span> <span class="n">term</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;/&#39;</span><span class="p">)[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">replaced_by_id</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">g</span>

<div class="viewcode-block" id="is_secondary"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_secondary">[docs]</a><span class="k">def</span> <span class="nf">is_secondary</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the UniProt ID corresponds to a secondary accession.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID to check.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if it is a secondary accessing entry, False otherwise.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">entry</span> <span class="o">=</span> <span class="n">uniprot_sec</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">entry</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">return</span> <span class="kc">True</span></div>

<div class="viewcode-block" id="get_primary_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_primary_id">[docs]</a><span class="k">def</span> <span class="nf">get_primary_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return a primary entry corresponding to the UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        The UniProt ID to map to primary.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    primary_id : str</span>
<span class="sd">        If the given ID is primary, it is returned as is. Othwewise the primary</span>
<span class="sd">        IDs are looked up. If there are multiple primary IDs then the first</span>
<span class="sd">        human one is returned. If there are no human primary IDs then the</span>
<span class="sd">        first primary found is returned.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">primaries</span> <span class="o">=</span> <span class="n">uniprot_sec</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">primaries</span><span class="p">:</span>
        <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">primaries</span><span class="p">)</span> <span class="o">&gt;</span> <span class="mi">1</span><span class="p">:</span>
            <span class="n">logger</span><span class="o">.</span><span class="n">debug</span><span class="p">(</span><span class="s1">&#39;More than 1 primary ID for </span><span class="si">%s</span><span class="s1">.&#39;</span> <span class="o">%</span> <span class="n">protein_id</span><span class="p">)</span>
            <span class="k">for</span> <span class="n">primary</span> <span class="ow">in</span> <span class="n">primaries</span><span class="p">:</span>
                <span class="c1"># Often secondary IDs were broken into multiple primary IDs</span>
                <span class="c1"># for different organisms. In this case we return the human</span>
                <span class="c1"># one if it exists.</span>
                <span class="k">if</span> <span class="n">is_human</span><span class="p">(</span><span class="n">primary</span><span class="p">):</span>
                    <span class="k">return</span> <span class="n">primary</span>
        <span class="c1"># If we haven&#39;t returned anything then we just return the</span>
        <span class="c1"># first primary id</span>
        <span class="k">return</span> <span class="n">primaries</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="c1"># If there is not secondary entry the we assume this is a primary entry</span>
    <span class="k">return</span> <span class="n">protein_id</span></div>

<div class="viewcode-block" id="get_family_members"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_family_members">[docs]</a><span class="k">def</span> <span class="nf">get_family_members</span><span class="p">(</span><span class="n">family_name</span><span class="p">,</span> <span class="n">human_only</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the HGNC gene symbols which are the members of a given family.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    family_name : str</span>
<span class="sd">        Family name to be queried.</span>
<span class="sd">    human_only : bool</span>
<span class="sd">        If True, only human proteins in the family will be returned.</span>
<span class="sd">        Default: True</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    gene_names : list</span>
<span class="sd">        The HGNC gene symbols corresponding to the given family.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">data</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;query&#39;</span><span class="p">:</span> <span class="s1">&#39;family:</span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">family_name</span><span class="p">,</span> 
            <span class="s1">&#39;format&#39;</span><span class="p">:</span> <span class="s1">&#39;list&#39;</span><span class="p">}</span>
    <span class="k">if</span> <span class="n">human_only</span><span class="p">:</span>
        <span class="n">data</span><span class="p">[</span><span class="s1">&#39;fil&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="s1">&#39;organism:human&#39;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">uniprot_url</span><span class="p">,</span> <span class="n">params</span><span class="o">=</span><span class="n">data</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">res</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span> <span class="ow">or</span> <span class="ow">not</span> <span class="n">res</span><span class="o">.</span><span class="n">text</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="c1"># res.text gets us the Unicode</span>
    <span class="n">html</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">text</span>
    <span class="n">protein_list</span> <span class="o">=</span> <span class="n">html</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="n">gene_names</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">p</span> <span class="ow">in</span> <span class="n">protein_list</span><span class="p">:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="n">get_gene_name</span><span class="p">(</span><span class="n">p</span><span class="p">)</span>
        <span class="n">gene_names</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">gene_name</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">gene_names</span></div>

<div class="viewcode-block" id="get_mnemonic"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_mnemonic">[docs]</a><span class="k">def</span> <span class="nf">get_mnemonic</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">web_fallback</span><span class="o">=</span><span class="kc">False</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt mnemonic for the given UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID to be mapped.</span>
<span class="sd">    web_fallback : Optional[bool]</span>
<span class="sd">        If True and the offline lookup fails, the UniProt web service</span>
<span class="sd">        is used to do the query.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mnemonic : str</span>
<span class="sd">        The UniProt mnemonic corresponding to the given Uniprot ID.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">mnemonic</span> <span class="o">=</span> <span class="n">uniprot_mnemonic</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">mnemonic</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">web_fallback</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?mnemonic</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?r up:mnemonic ?mnemonic .</span>
<span class="s2">        }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">mnemonic</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">toPython</span><span class="p">()</span>
        <span class="k">return</span> <span class="n">mnemonic</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="get_id_from_mnemonic"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_id_from_mnemonic">[docs]</a><span class="k">def</span> <span class="nf">get_id_from_mnemonic</span><span class="p">(</span><span class="n">uniprot_mnemonic</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt ID for the given UniProt mnemonic.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    uniprot_mnemonic : str</span>
<span class="sd">        UniProt mnemonic to be mapped.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    uniprot_id : str</span>
<span class="sd">        The UniProt ID corresponding to the given Uniprot mnemonic.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">uniprot_id</span> <span class="o">=</span> <span class="n">uniprot_mnemonic_reverse</span><span class="p">[</span><span class="n">uniprot_mnemonic</span><span class="p">]</span>
        <span class="k">return</span> <span class="n">uniprot_id</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span></div>

<div class="viewcode-block" id="get_gene_name"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_gene_name">[docs]</a><span class="k">def</span> <span class="nf">get_gene_name</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">web_fallback</span><span class="o">=</span><span class="kc">True</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the gene name for the given UniProt ID.</span>

<span class="sd">    This is an alternative to get_hgnc_name and is useful when</span>
<span class="sd">    HGNC name is not availabe (for instance, when the organism</span>
<span class="sd">    is not homo sapiens).</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID to be mapped.</span>
<span class="sd">    web_fallback : Optional[bool]</span>
<span class="sd">        If True and the offline lookup fails, the UniProt web service</span>
<span class="sd">        is used to do the query.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    gene_name : str</span>
<span class="sd">        The gene name corresponding to the given Uniprot ID.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">protein_id</span> <span class="o">=</span> <span class="n">get_primary_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="n">uniprot_gene_name</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="c1"># There are cases when the entry is in the resource</span>
        <span class="c1"># table but the gene name is empty. Often this gene</span>
        <span class="c1"># name is actually available in the web service RDF</span>
        <span class="c1"># so here we return only if the gene name is not None</span>
        <span class="c1"># and not empty string.</span>
        <span class="k">if</span> <span class="n">gene_name</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">gene_name</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">web_fallback</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>

    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?name</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?gene a up:Gene .</span>
<span class="s2">            ?gene skos:prefLabel ?name .</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">gene_name</span> <span class="o">=</span> <span class="p">[</span><span class="n">r</span> <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">][</span><span class="mi">0</span><span class="p">][</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">toPython</span><span class="p">()</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">gene_name</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">None</span>
        <span class="k">return</span> <span class="n">gene_name</span>
    <span class="k">return</span> <span class="kc">None</span></div>

<span class="nd">@lru_cache</span><span class="p">(</span><span class="n">maxsize</span><span class="o">=</span><span class="mi">1000</span><span class="p">)</span>
<span class="k">def</span> <span class="nf">get_sequence</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">prim_ids</span> <span class="o">=</span> <span class="n">uniprot_sec</span><span class="p">[</span><span class="n">protein_id</span><span class="p">]</span>
        <span class="n">protein_id</span> <span class="o">=</span> <span class="n">prim_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="n">url</span> <span class="o">=</span> <span class="n">uniprot_url</span> <span class="o">+</span> <span class="s1">&#39;</span><span class="si">%s</span><span class="s1">.fasta&#39;</span> <span class="o">%</span> <span class="n">protein_id</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">requests</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">url</span><span class="p">)</span>
    <span class="k">if</span> <span class="ow">not</span> <span class="n">res</span><span class="o">.</span><span class="n">status_code</span> <span class="o">==</span> <span class="mi">200</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Could not find sequence for protein </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">protein_id</span><span class="p">)</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="c1"># res.text is Unicode</span>
    <span class="n">lines</span> <span class="o">=</span> <span class="n">res</span><span class="o">.</span><span class="n">text</span><span class="o">.</span><span class="n">splitlines</span><span class="p">()</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="p">(</span><span class="s1">&#39;&#39;</span><span class="o">.</span><span class="n">join</span><span class="p">(</span><span class="n">lines</span><span class="p">[</span><span class="mi">1</span><span class="p">:]))</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="s1">&#39;</span><span class="se">\n</span><span class="s1">&#39;</span><span class="p">,</span><span class="s1">&#39;&#39;</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">seq</span>


<span class="k">def</span> <span class="nf">get_modifications</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="n">g</span> <span class="o">=</span> <span class="n">query_protein</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">g</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">None</span>
    <span class="n">query</span> <span class="o">=</span> <span class="n">rdf_prefixes</span> <span class="o">+</span> <span class="s2">&quot;&quot;&quot;</span>
<span class="s2">        SELECT ?beg_pos ?comment</span>
<span class="s2">        WHERE {</span>
<span class="s2">            ?mod_res a up:Modified_Residue_Annotation .</span>
<span class="s2">            ?mod_res rdfs:comment ?comment .</span>
<span class="s2">            ?mod_res up:range ?range .</span>
<span class="s2">            ?range faldo:begin ?beg .</span>
<span class="s2">            ?range faldo:end ?end .</span>
<span class="s2">            ?beg a faldo:ExactPosition .</span>
<span class="s2">            ?beg faldo:position ?beg_pos .</span>
<span class="s2">            FILTER (?beg = ?end)</span>
<span class="s2">            }</span>
<span class="s2">        &quot;&quot;&quot;</span>
    <span class="n">res</span> <span class="o">=</span> <span class="n">g</span><span class="o">.</span><span class="n">query</span><span class="p">(</span><span class="n">query</span><span class="p">)</span>
    <span class="n">mods</span> <span class="o">=</span> <span class="p">[]</span>
    <span class="k">for</span> <span class="n">r</span> <span class="ow">in</span> <span class="n">res</span><span class="p">:</span>
        <span class="n">mod_pos</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">.</span><span class="n">value</span>
        <span class="c1"># &quot;Phosphothreonine; by autocatalysis&quot;</span>
        <span class="c1"># &quot;Phosphothreonine; by MAP2K1 and MAP2K2&quot;</span>
        <span class="c1"># TODO: take into account the comment after the ;?</span>
        <span class="n">mod_res</span> <span class="o">=</span> <span class="n">r</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
        <span class="n">mods</span><span class="o">.</span><span class="n">append</span><span class="p">((</span><span class="n">mod_res</span><span class="p">,</span> <span class="n">mod_pos</span><span class="p">))</span>
    <span class="k">return</span> <span class="n">mods</span>


<div class="viewcode-block" id="verify_location"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.verify_location">[docs]</a><span class="k">def</span> <span class="nf">verify_location</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">residue</span><span class="p">,</span> <span class="n">location</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the residue is at the given location in the UP sequence.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein whose sequence is used as reference.</span>
<span class="sd">    residue : str</span>
<span class="sd">        A single character amino acid symbol (Y, S, T, V, etc.)</span>
<span class="sd">    location : str</span>
<span class="sd">        The location on the protein sequence (starting at 1) at which the</span>
<span class="sd">        residue should be checked against the reference sequence.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the given residue is at the given position in the sequence</span>
<span class="sd">    corresponding to the given UniProt ID, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">get_sequence</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="c1"># If we couldn&#39;t get the sequence (can happen due to web service hiccups)</span>
    <span class="c1"># don&#39;t throw the statement away by default</span>
    <span class="k">if</span> <span class="n">seq</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">loc_int</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">location</span><span class="p">)</span>
    <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
        <span class="n">logger</span><span class="o">.</span><span class="n">warning</span><span class="p">(</span><span class="s1">&#39;Invalid location </span><span class="si">%s</span><span class="s1">&#39;</span> <span class="o">%</span> <span class="n">location</span><span class="p">)</span>
        <span class="n">loc_int</span> <span class="o">=</span> <span class="o">-</span><span class="mi">1</span>

    <span class="k">if</span> <span class="p">(</span><span class="n">loc_int</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">)</span> <span class="ow">or</span> <span class="p">(</span><span class="n">loc_int</span> <span class="o">&gt;</span> <span class="nb">len</span><span class="p">(</span><span class="n">seq</span><span class="p">)):</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">elif</span> <span class="n">seq</span><span class="p">[</span><span class="n">loc_int</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">residue</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="kc">False</span></div>


<div class="viewcode-block" id="verify_modification"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.verify_modification">[docs]</a><span class="k">def</span> <span class="nf">verify_modification</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">residue</span><span class="p">,</span> <span class="n">location</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the residue at the given location has a known modifiation. </span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein whose sequence is used as reference.</span>
<span class="sd">    residue : str</span>
<span class="sd">        A single character amino acid symbol (Y, S, T, V, etc.)</span>
<span class="sd">    location : Optional[str]</span>
<span class="sd">        The location on the protein sequence (starting at 1) at which the</span>
<span class="sd">        modification is checked.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the given residue is reported to be modified at the given position </span>
<span class="sd">    in the sequence corresponding to the given UniProt ID, otherwise False.</span>
<span class="sd">    If location is not given, we only check if there is any residue of the</span>
<span class="sd">    given type that is modified.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">mods</span> <span class="o">=</span> <span class="n">get_modifications</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="n">mod_locs</span> <span class="o">=</span> <span class="p">[</span><span class="n">m</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="k">for</span> <span class="n">m</span> <span class="ow">in</span> <span class="n">mods</span><span class="p">]</span>
    <span class="n">seq</span> <span class="o">=</span> <span class="n">get_sequence</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">location</span><span class="p">:</span>
        <span class="k">if</span> <span class="ow">not</span> <span class="n">verify_location</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">residue</span><span class="p">,</span> <span class="n">location</span><span class="p">):</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">try</span><span class="p">:</span>
            <span class="n">mod_idx</span> <span class="o">=</span> <span class="n">mod_locs</span><span class="o">.</span><span class="n">index</span><span class="p">(</span><span class="n">location</span><span class="p">)</span>
        <span class="k">except</span> <span class="ne">ValueError</span><span class="p">:</span>
            <span class="k">return</span> <span class="kc">False</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">for</span> <span class="n">ml</span> <span class="ow">in</span> <span class="n">mod_locs</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">seq</span><span class="p">[</span><span class="n">ml</span> <span class="o">-</span> <span class="mi">1</span><span class="p">]</span> <span class="o">==</span> <span class="n">residue</span><span class="p">:</span>
                <span class="k">return</span> <span class="kc">True</span>
        <span class="k">return</span> <span class="kc">False</span></div>

<span class="k">def</span> <span class="nf">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="n">organism_suffix</span><span class="p">):</span>
    <span class="n">mnemonic</span> <span class="o">=</span> <span class="n">get_mnemonic</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">mnemonic</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="k">return</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="n">mnemonic</span><span class="o">.</span><span class="n">endswith</span><span class="p">(</span><span class="n">organism_suffix</span><span class="p">):</span>
        <span class="k">return</span> <span class="kc">True</span>
    <span class="k">return</span> <span class="kc">False</span>

<div class="viewcode-block" id="is_human"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_human">[docs]</a><span class="k">def</span> <span class="nf">is_human</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the given protein id corresponds to a human protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the protein_id corresponds to a human protein, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="s1">&#39;HUMAN&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="is_mouse"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_mouse">[docs]</a><span class="k">def</span> <span class="nf">is_mouse</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the given protein id corresponds to a mouse protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the protein_id corresponds to a mouse protein, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="s1">&#39;MOUSE&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="is_rat"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.is_rat">[docs]</a><span class="k">def</span> <span class="nf">is_rat</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return True if the given protein id corresponds to a rat protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    True if the protein_id corresponds to a rat protein, otherwise False.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">_is_organism</span><span class="p">(</span><span class="n">protein_id</span><span class="p">,</span> <span class="s1">&#39;RAT&#39;</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_mgi_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_mgi_id">[docs]</a><span class="k">def</span> <span class="nf">get_mgi_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the MGI ID given the protein id of a mouse protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the mouse protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mgi_id : str</span>
<span class="sd">        MGI ID of the mouse protein</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">uniprot_mgi</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_rgd_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_rgd_id">[docs]</a><span class="k">def</span> <span class="nf">get_rgd_id</span><span class="p">(</span><span class="n">protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the RGD ID given the protein id of a rat protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    protein_id : str</span>
<span class="sd">        UniProt ID of the rat protein</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    rgd_id : str</span>
<span class="sd">        RGD ID of the rat protein</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">uniprot_rgd</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">protein_id</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_id_from_mgi"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_id_from_mgi">[docs]</a><span class="k">def</span> <span class="nf">get_id_from_mgi</span><span class="p">(</span><span class="n">mgi_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt ID given the MGI ID of a mouse protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    mgi_id : str</span>
<span class="sd">        The MGI ID of the mouse protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    up_id : str</span>
<span class="sd">        The UniProt ID of the mouse protein.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">uniprot_mgi_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mgi_id</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_id_from_rgd"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_id_from_rgd">[docs]</a><span class="k">def</span> <span class="nf">get_id_from_rgd</span><span class="p">(</span><span class="n">rgd_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the UniProt ID given the RGD ID of a rat protein.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    rgd_id : str</span>
<span class="sd">        The RGD ID of the rat protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    up_id : str</span>
<span class="sd">        The UniProt ID of the rat protein.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">uniprot_rgd_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rgd_id</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_mouse_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_mouse_id">[docs]</a><span class="k">def</span> <span class="nf">get_mouse_id</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the mouse UniProt ID given a human UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    human_protein_id : str</span>
<span class="sd">        The UniProt ID of a human protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    mouse_protein_id : str</span>
<span class="sd">        The UniProt ID of a mouse protein orthologous to the given human protein</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">uniprot_human_mouse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">)</span></div>

<div class="viewcode-block" id="get_rat_id"><a class="viewcode-back" href="../../../modules/databases/index.html#indra.databases.uniprot_client.get_rat_id">[docs]</a><span class="k">def</span> <span class="nf">get_rat_id</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">):</span>
    <span class="sd">&quot;&quot;&quot;Return the rat UniProt ID given a human UniProt ID.</span>

<span class="sd">    Parameters</span>
<span class="sd">    ----------</span>
<span class="sd">    human_protein_id : str</span>
<span class="sd">        The UniProt ID of a human protein.</span>

<span class="sd">    Returns</span>
<span class="sd">    -------</span>
<span class="sd">    rat_protein_id : str</span>
<span class="sd">        The UniProt ID of a rat protein orthologous to the given human protein</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="k">return</span> <span class="n">uniprot_human_rat</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">human_protein_id</span><span class="p">)</span></div>

<span class="k">def</span> <span class="nf">_build_uniprot_entries</span><span class="p">():</span>
    <span class="n">up_entries_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span> \
        <span class="s1">&#39;/../resources/uniprot_entries.tsv&#39;</span>
    <span class="n">uniprot_gene_name</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mnemonic</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mnemonic_reverse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mgi</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_rgd</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_mgi_reverse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_rgd_reverse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">csv_rows</span> <span class="o">=</span> <span class="n">read_unicode_csv</span><span class="p">(</span><span class="n">up_entries_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1"># Skip the header row</span>
        <span class="nb">next</span><span class="p">(</span><span class="n">csv_rows</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_rows</span><span class="p">:</span>
            <span class="n">up_id</span><span class="p">,</span> <span class="n">gene_name</span><span class="p">,</span> <span class="n">up_mnemonic</span><span class="p">,</span> <span class="n">rgd</span><span class="p">,</span> <span class="n">mgi</span> <span class="o">=</span> <span class="n">row</span>
            <span class="n">uniprot_gene_name</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">gene_name</span>
            <span class="n">uniprot_mnemonic</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">up_mnemonic</span>
            <span class="n">uniprot_mnemonic_reverse</span><span class="p">[</span><span class="n">up_mnemonic</span><span class="p">]</span> <span class="o">=</span> <span class="n">up_id</span>
            <span class="k">if</span> <span class="n">mgi</span><span class="p">:</span>
                <span class="n">mgi_ids</span> <span class="o">=</span> <span class="n">mgi</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">mgi_ids</span><span class="p">:</span>
                    <span class="n">uniprot_mgi</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">mgi_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">uniprot_mgi_reverse</span><span class="p">[</span><span class="n">mgi_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">up_id</span>
            <span class="k">if</span> <span class="n">rgd</span><span class="p">:</span>
                <span class="n">rgd_ids</span> <span class="o">=</span> <span class="n">rgd</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;;&#39;</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">rgd_ids</span><span class="p">:</span>
                    <span class="n">uniprot_rgd</span><span class="p">[</span><span class="n">up_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">rgd_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
                    <span class="n">uniprot_rgd_reverse</span><span class="p">[</span><span class="n">rgd_ids</span><span class="p">[</span><span class="mi">0</span><span class="p">]]</span> <span class="o">=</span> <span class="n">up_id</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="k">return</span> <span class="p">(</span><span class="n">uniprot_gene_name</span><span class="p">,</span> <span class="n">uniprot_mnemonic</span><span class="p">,</span> <span class="n">uniprot_mnemonic_reverse</span><span class="p">,</span> \
            <span class="n">uniprot_mgi</span><span class="p">,</span> <span class="n">uniprot_rgd</span><span class="p">,</span> <span class="n">uniprot_mgi_reverse</span><span class="p">,</span> <span class="n">uniprot_rgd_reverse</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">_build_human_mouse_rat</span><span class="p">():</span>
    <span class="n">hgnc_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span>\
                <span class="s1">&#39;/../resources/hgnc_entries.tsv&#39;</span>
    <span class="n">csv_rows</span> <span class="o">=</span> <span class="n">read_unicode_csv</span><span class="p">(</span><span class="n">hgnc_file</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
    <span class="c1"># Skip the header row</span>
    <span class="nb">next</span><span class="p">(</span><span class="n">csv_rows</span><span class="p">)</span>
    <span class="n">uniprot_mouse</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="n">uniprot_rat</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_rows</span><span class="p">:</span>
        <span class="n">human_id</span><span class="p">,</span> <span class="n">mgi_id</span><span class="p">,</span> <span class="n">rgd_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">6</span><span class="p">:</span><span class="mi">9</span><span class="p">]</span>
        <span class="k">if</span> <span class="n">human_id</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">mgi_id</span><span class="p">:</span>
                <span class="n">mgi_id</span> <span class="o">=</span> <span class="n">mgi_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">mgi_id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;MGI:&#39;</span><span class="p">):</span>
                    <span class="n">mgi_id</span> <span class="o">=</span> <span class="n">mgi_id</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>
                <span class="n">mouse_id</span> <span class="o">=</span> <span class="n">uniprot_mgi_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">mgi_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">mouse_id</span><span class="p">:</span>
                    <span class="n">uniprot_mouse</span><span class="p">[</span><span class="n">human_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">mouse_id</span>
            <span class="k">if</span> <span class="n">rgd_id</span><span class="p">:</span>
                <span class="n">rgd_id</span> <span class="o">=</span> <span class="n">rgd_id</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;, &#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">rgd_id</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;RGD:&#39;</span><span class="p">):</span>
                    <span class="n">rgd_id</span> <span class="o">=</span> <span class="n">rgd_id</span><span class="p">[</span><span class="mi">4</span><span class="p">:]</span>
                <span class="n">rat_id</span> <span class="o">=</span> <span class="n">uniprot_rgd_reverse</span><span class="o">.</span><span class="n">get</span><span class="p">(</span><span class="n">rgd_id</span><span class="p">)</span>
                <span class="k">if</span> <span class="n">rat_id</span><span class="p">:</span>
                    <span class="n">uniprot_rat</span><span class="p">[</span><span class="n">human_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">rat_id</span>
    <span class="k">return</span> <span class="n">uniprot_mouse</span><span class="p">,</span> <span class="n">uniprot_rat</span>

<span class="k">def</span> <span class="nf">_build_uniprot_sec</span><span class="p">():</span>
    <span class="c1"># File containing secondary accession numbers mapped</span>
    <span class="c1"># to primary accession numbers</span>
    <span class="n">sec_file</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span>\
                <span class="s1">&#39;/../resources/uniprot_sec_ac.txt&#39;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">uniprot_sec</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="n">lines</span> <span class="o">=</span> <span class="nb">open</span><span class="p">(</span><span class="n">sec_file</span><span class="p">,</span> <span class="s1">&#39;rt&#39;</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">l</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">lines</span><span class="p">):</span>
            <span class="k">if</span> <span class="n">l</span><span class="o">.</span><span class="n">startswith</span><span class="p">(</span><span class="s1">&#39;Secondary AC&#39;</span><span class="p">):</span>
                <span class="n">entry_lines</span> <span class="o">=</span> <span class="n">lines</span><span class="p">[</span><span class="n">i</span><span class="o">+</span><span class="mi">2</span><span class="p">:]</span>

        <span class="k">for</span> <span class="n">l</span> <span class="ow">in</span> <span class="n">entry_lines</span><span class="p">:</span>
            <span class="n">sec_id</span><span class="p">,</span> <span class="n">prim_id</span> <span class="o">=</span> <span class="n">l</span><span class="o">.</span><span class="n">split</span><span class="p">()</span>
            <span class="k">try</span><span class="p">:</span>
                <span class="n">uniprot_sec</span><span class="p">[</span><span class="n">sec_id</span><span class="p">]</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">prim_id</span><span class="p">)</span>
            <span class="k">except</span> <span class="ne">KeyError</span><span class="p">:</span>
                <span class="n">uniprot_sec</span><span class="p">[</span><span class="n">sec_id</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">prim_id</span><span class="p">]</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">uniprot_sec</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">return</span> <span class="n">uniprot_sec</span>

<span class="k">def</span> <span class="nf">_build_uniprot_subcell_loc</span><span class="p">():</span>
    <span class="n">fname</span> <span class="o">=</span> <span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">dirname</span><span class="p">(</span><span class="n">os</span><span class="o">.</span><span class="n">path</span><span class="o">.</span><span class="n">abspath</span><span class="p">(</span><span class="vm">__file__</span><span class="p">))</span> <span class="o">+</span>\
                <span class="s1">&#39;/../resources/uniprot_subcell_loc.tsv&#39;</span>
    <span class="k">try</span><span class="p">:</span>
        <span class="n">csv_rows</span> <span class="o">=</span> <span class="n">read_unicode_csv</span><span class="p">(</span><span class="n">fname</span><span class="p">,</span> <span class="n">delimiter</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\t</span><span class="s1">&#39;</span><span class="p">)</span>
        <span class="c1"># Skip the header row</span>
        <span class="nb">next</span><span class="p">(</span><span class="n">csv_rows</span><span class="p">)</span>
        <span class="n">subcell_loc</span> <span class="o">=</span> <span class="p">{}</span>
        <span class="k">for</span> <span class="n">row</span> <span class="ow">in</span> <span class="n">csv_rows</span><span class="p">:</span>
            <span class="n">loc_id</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">loc_alias</span> <span class="o">=</span> <span class="n">row</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>
            <span class="n">subcell_loc</span><span class="p">[</span><span class="n">loc_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">loc_alias</span>
    <span class="k">except</span> <span class="ne">IOError</span><span class="p">:</span>
        <span class="n">subcell_loc</span> <span class="o">=</span> <span class="p">{}</span>
    <span class="k">return</span> <span class="n">subcell_loc</span>

<span class="p">(</span><span class="n">uniprot_gene_name</span><span class="p">,</span> <span class="n">uniprot_mnemonic</span><span class="p">,</span> <span class="n">uniprot_mnemonic_reverse</span><span class="p">,</span>
 <span class="n">uniprot_mgi</span><span class="p">,</span> <span class="n">uniprot_rgd</span><span class="p">,</span> <span class="n">uniprot_mgi_reverse</span><span class="p">,</span> <span class="n">uniprot_rgd_reverse</span><span class="p">)</span> <span class="o">=</span> \
 <span class="n">_build_uniprot_entries</span><span class="p">()</span>
<span class="n">uniprot_sec</span> <span class="o">=</span> <span class="n">_build_uniprot_sec</span><span class="p">()</span>
<span class="n">uniprot_subcell_loc</span> <span class="o">=</span> <span class="n">_build_uniprot_subcell_loc</span><span class="p">()</span>
<span class="n">uniprot_human_mouse</span><span class="p">,</span> <span class="n">uniprot_human_rat</span> <span class="o">=</span> <span class="n">_build_human_mouse_rat</span><span class="p">()</span>
</pre></div>

          </div>
          <footer>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, B. M. Gyori, J. A. Bachman.
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../../',
            VERSION:'1.5.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>