

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tools (indra.tools) &mdash; INDRA 1.5.0 documentation</title>
  

  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="../../_static/css/theme.css" type="text/css" />
  

  

  
    <link rel="top" title="INDRA 1.5.0 documentation" href="../../index.html"/>
        <link rel="up" title="INDRA modules reference" href="../index.html"/>
        <link rel="next" title="Tutorials" href="../../tutorials/index.html"/>
        <link rel="prev" title="Explanation (indra.explanation)" href="../explanation/index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-nav-search">
        

        
          <a href="../../index.html" class="icon icon-home"> INDRA
        

        
        </a>

        
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

        
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
        
          
          
              <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../../installation.html">Installation</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#installing-python">Installing Python</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#installing-indra">Installing INDRA</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#installing-via-github">Installing via Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#cloning-the-source-code-from-github">Cloning the source code from Github</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#installing-releases-with-pip">Installing releases with pip</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../installation.html#indra-dependencies">INDRA dependencies</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#pysb-and-bionetgen">PySB and BioNetGen</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#pyjnius">Pyjnius</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#graphviz">Graphviz</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#matplotlib">Matplotlib</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../installation.html#optional-additional-dependencies">Optional additional dependencies</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../getting_started.html">Getting started with INDRA</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../getting_started.html#importing-indra-and-its-modules">Importing INDRA and its modules</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../getting_started.html#basic-usage-examples">Basic usage examples</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../getting_started.html#reading-a-sentence-with-trips">Reading a sentence with TRIPS</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../getting_started.html#reading-a-pubmed-central-article-with-reach">Reading a PubMed Central article with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../getting_started.html#getting-the-neighborhood-of-proteins-from-the-bel-large-corpus">Getting the neighborhood of proteins from the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../getting_started.html#getting-paths-between-two-proteins-from-pathwaycommons-biopax">Getting paths between two proteins from PathwayCommons (BioPAX)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../getting_started.html#constructing-indra-statements-manually">Constructing INDRA Statements manually</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../getting_started.html#assembling-a-pysb-model-and-exporting-to-sbml">Assembling a PySB model and exporting to SBML</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1 current"><a class="reference internal" href="../index.html">INDRA modules reference</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="../statements.html">INDRA Statements (<code class="docutils literal"><span class="pre">indra.statements</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../sources/index.html">Processors for model input (<code class="docutils literal"><span class="pre">indra.sources</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../sources/bel/index.html">BEL (<code class="docutils literal"><span class="pre">indra.sources.bel</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../sources/bel/index.html#module-indra.sources.bel.bel_api">BEL API (<code class="docutils literal"><span class="pre">indra.sources.bel.bel_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sources/bel/index.html#module-indra.sources.bel.processor">BEL Processor (<code class="docutils literal"><span class="pre">indra.sources.bel.processor</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../sources/biopax/index.html">Biopax (<code class="docutils literal"><span class="pre">indra.sources.biopax</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../sources/biopax/index.html#module-indra.sources.biopax.biopax_api">Biopax API (<code class="docutils literal"><span class="pre">indra.sources.biopax.biopax_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sources/biopax/index.html#module-indra.sources.biopax.processor">Biopax Processor (<code class="docutils literal"><span class="pre">indra.sources.biopax.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sources/biopax/index.html#module-indra.sources.biopax.pathway_commons_client">Pathway Commons Client (<code class="docutils literal"><span class="pre">indra.sources.biopax.pathway_commons_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../sources/reach/index.html">REACH (<code class="docutils literal"><span class="pre">indra.sources.reach</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../sources/reach/index.html#module-indra.sources.reach.reach_api">REACH API (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sources/reach/index.html#module-indra.sources.reach.processor">REACH Processor (<code class="docutils literal"><span class="pre">indra.sources.reach.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sources/reach/index.html#module-indra.sources.reach.reach_reader">REACH reader (<code class="docutils literal"><span class="pre">indra.sources.reach.reach_reader</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../sources/trips/index.html">TRIPS (<code class="docutils literal"><span class="pre">indra.sources.trips</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../sources/trips/index.html#module-indra.sources.trips.trips_api">TRIPS API (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_api</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sources/trips/index.html#module-indra.sources.trips.processor">TRIPS Processor (<code class="docutils literal"><span class="pre">indra.sources.trips.processor</span></code>)</a></li>
<li class="toctree-l4"><a class="reference internal" href="../sources/trips/index.html#module-indra.sources.trips.trips_client">TRIPS Client (<code class="docutils literal"><span class="pre">indra.sources.trips.trips_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l3"><a class="reference internal" href="../sources/ndex_cx/index.html">NDEx CX API (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx_api</span></code>)</a><ul>
<li class="toctree-l4"><a class="reference internal" href="../sources/ndex_cx/index.html#ndex-cx-processor-indra-sources-ndex-cx-processor">NDEx CX Processor (<code class="docutils literal"><span class="pre">indra.sources.ndex_cx.processor</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../databases/index.html">Database clients (<code class="docutils literal"><span class="pre">indra.databases</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.hgnc_client">HGNC client (<code class="docutils literal"><span class="pre">indra.hgnc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.uniprot_client">Uniprot client (<code class="docutils literal"><span class="pre">indra.databases.uniprot_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.chebi_client">ChEBI client (<code class="docutils literal"><span class="pre">indra.databases.chebi_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.biogrid_client">BioGRID client (<code class="docutils literal"><span class="pre">indra.databases.biogrid_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.context_client">Cell type context client (<code class="docutils literal"><span class="pre">indra.databases.context_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.relevance_client">Network relevance client (<code class="docutils literal"><span class="pre">indra.databases.relevance_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.ndex_client">NDEx client (<code class="docutils literal"><span class="pre">indra.databases.ndex_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../databases/index.html#module-indra.databases.cbio_client">cBio portal client (<code class="docutils literal"><span class="pre">indra.databases.cbio_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../literature/index.html">Literature clients (<code class="docutils literal"><span class="pre">indra.literature</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../literature/index.html#module-indra.literature.pubmed_client">Pubmed client (<code class="docutils literal"><span class="pre">indra.literature.pubmed_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../literature/index.html#module-indra.literature.pmc_client">Pubmed Central client (<code class="docutils literal"><span class="pre">indra.literature.pmc_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../literature/index.html#module-indra.literature.crossref_client">CrossRef client (<code class="docutils literal"><span class="pre">indra.literature.crossref_client</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../literature/index.html#module-indra.literature.elsevier_client">Elsevier client (<code class="docutils literal"><span class="pre">indra.literature.elsevier_client</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../preassembler/index.html">Preassembly (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../preassembler/index.html#module-indra.preassembler">Preassembler (<code class="docutils literal"><span class="pre">indra.preassembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../preassembler/index.html#module-indra.preassembler.grounding_mapper">Entity grounding curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.grounding_mapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../preassembler/index.html#module-indra.preassembler.sitemapper">Site curation and mapping (<code class="docutils literal"><span class="pre">indra.preassembler.sitemapper</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../preassembler/index.html#module-indra.preassembler.hierarchy_manager">Hierarchy manager (<code class="docutils literal"><span class="pre">indra.preassembler.hierarchy_manager</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../belief/index.html">Belief Engine (<code class="docutils literal"><span class="pre">indra.belief</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../mechlinker/index.html">Mechanism Linker (<code class="docutils literal"><span class="pre">indra.mechlinker</span></code>)</a></li>
<li class="toctree-l2"><a class="reference internal" href="../assemblers/index.html">Assemblers of model output (<code class="docutils literal"><span class="pre">indra.assemblers</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../assemblers/pysb_assembler.html">Executable PySB models (<code class="docutils literal"><span class="pre">indra.assemblers.pysb_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../assemblers/cx_assembler.html">Cytoscape networks (<code class="docutils literal"><span class="pre">indra.assemblers.cx_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../assemblers/english_assembler.html">Natural language (<code class="docutils literal"><span class="pre">indra.assemblers.english_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../assemblers/graph_assembler.html">Node-edge graphs (<code class="docutils literal"><span class="pre">indra.assemblers.graph_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../assemblers/sif_assembler.html">SIF / Boolean networks (<code class="docutils literal"><span class="pre">indra.assemblers.sif_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../assemblers/index_card_assembler.html">MITRE &#8220;index cards&#8221; (<code class="docutils literal"><span class="pre">indra.assemblers.index_card_assembler</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="../assemblers/sbgn_assembler.html">SBGN output (<code class="docutils literal"><span class="pre">indra.assemblers.sbgn_assembler</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../explanation/index.html">Explanation (<code class="docutils literal"><span class="pre">indra.explanation</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../explanation/index.html#module-indra.explanation.model_checker">Check whether a rule-based model satisfies a property (<code class="docutils literal"><span class="pre">indra.explanation.model_checker</span></code>)</a></li>
</ul>
</li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Tools (<code class="docutils literal"><span class="pre">indra.tools</span></code>)</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#module-indra.tools.assemble_corpus">Run assembly components in a pipeline (<code class="docutils literal"><span class="pre">indra.tools.assemble_corpus</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-indra.tools.gene_network">Build a network from a gene list (<code class="docutils literal"><span class="pre">indra.tools.gene_network</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-indra.tools.executable_subnetwork">Build an executable model from a fragment of a large network (<code class="docutils literal"><span class="pre">indra.tools.executable_subnetwork</span></code>)</a></li>
<li class="toctree-l3"><a class="reference internal" href="#module-indra.tools.incremental_model">Build a model incrementally over time (<code class="docutils literal"><span class="pre">indra.tools.incremental_model</span></code>)</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../tutorials/index.html">Tutorials</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/nl_modeling.html">Using natural language to build models</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/nl_modeling.html#read-indra-statements-from-a-natural-language-string">Read INDRA Statements from a natural language string</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/nl_modeling.html#assemble-the-indra-statements-into-a-rule-based-executable-model">Assemble the INDRA Statements into a rule-based executable model</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/nl_modeling.html#exporting-the-model-into-other-common-formats">Exporting the model into other common formats</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/machine_reading.html">Large-Scale Machine Reading with Starcluster</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading.html#install-reach">Install REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading.html#install-amazon-s3-support">Install Amazon S3 support</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading.html#install-other-dependencies">Install other dependencies</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading.html#assemble-a-corpus-of-pmids">Assemble a Corpus of PMIDs</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading.html#process-the-papers-with-reach">Process the papers with REACH</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading.html#extract-indra-statements-from-the-reach-output-on-s3">Extract INDRA Statements from the REACH output on S3</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading.html#running-the-whole-pipeline-with-one-script">Running the whole pipeline with one script</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/machine_reading_aws.html">Large-Scale Machine Reading with Amazon Batch</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/machine_reading_aws.html#how-it-works">How it Works</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="../../tutorials/gene_network.html">Assembling everything known about a particular gene</a><ul>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/gene_network.html#collect-mechanisms-from-pathwaycommons-and-the-bel-large-corpus">Collect mechanisms from PathwayCommons and the BEL Large Corpus</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/gene_network.html#collect-a-list-of-publications-that-discuss-the-gene-of-interest">Collect a list of publications that discuss the gene of interest</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/gene_network.html#get-the-full-text-or-abstract-corresponding-to-the-publications">Get the full text or abstract corresponding to the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/gene_network.html#read-the-content-of-the-publications">Read the content of the publications</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/gene_network.html#combine-all-statements-and-run-pre-assembly">Combine all statements and run pre-assembly</a></li>
<li class="toctree-l3"><a class="reference internal" href="../../tutorials/gene_network.html#assemble-the-statements-into-a-network-model">Assemble the statements into a network model</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../../rest_api.html">REST API</a><ul>
<li class="toctree-l2"><a class="reference internal" href="../../rest_api.html#installation">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../rest_api.html#launching-the-rest-service">Launching the REST service</a></li>
<li class="toctree-l2"><a class="reference internal" href="../../rest_api.html#documentation">Documentation</a></li>
</ul>
</li>
</ul>

          
        
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../../index.html">INDRA</a>
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../../index.html">Docs</a> &raquo;</li>
      
          <li><a href="../index.html">INDRA modules reference</a> &raquo;</li>
      
    <li>Tools (<code class="docutils literal"><span class="pre">indra.tools</span></code>)</li>
      <li class="wy-breadcrumbs-aside">
        
          
            <a href="../../_sources/modules/tools/index.txt" rel="nofollow"> View page source</a>
          
        
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document">
            
  <div class="section" id="tools-indra-tools">
<h1>Tools (<code class="xref py py-mod docutils literal"><span class="pre">indra.tools</span></code>)<a class="headerlink" href="#tools-indra-tools" title="Permalink to this headline">¶</a></h1>
<div class="section" id="module-indra.tools.assemble_corpus">
<span id="run-assembly-components-in-a-pipeline-indra-tools-assemble-corpus"></span><h2>Run assembly components in a pipeline (<a class="reference internal" href="#module-indra.tools.assemble_corpus" title="indra.tools.assemble_corpus"><code class="xref py py-mod docutils literal"><span class="pre">indra.tools.assemble_corpus</span></code></a>)<a class="headerlink" href="#module-indra.tools.assemble_corpus" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="indra.tools.assemble_corpus.dump_statements">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">dump_statements</code><span class="sig-paren">(</span><em>stmts</em>, <em>fname</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#dump_statements"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.dump_statements" title="Permalink to this definition">¶</a></dt>
<dd><p>Dump a list of statements into a pickle file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>fname</strong> (<em>str</em>) &#8211; The name of the pickle file to dump statements into.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.dump_stmt_strings">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">dump_stmt_strings</code><span class="sig-paren">(</span><em>stmts</em>, <em>fname</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#dump_stmt_strings"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.dump_stmt_strings" title="Permalink to this definition">¶</a></dt>
<dd><p>Save printed statements in a file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to save in a text file.</li>
<li><strong>fname</strong> (<em>Optional[str]</em>) &#8211; The name of a text file to save the printed statements into.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.expand_families">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">expand_families</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#expand_families"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.expand_families" title="Permalink to this definition">¶</a></dt>
<dd><p>Expand Bioentities Agents to individual genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to expand.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of expanded statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_belief">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_belief</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>belief_cutoff</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_belief"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_belief" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to statements with belief above a given cutoff.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>belief_cutoff</strong> (<em>float</em>) &#8211; Only statements with belief above the belief_cutoff will be returned.
Here 0 &lt; belief_cutoff &lt; 1.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_by_type">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_by_type</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>stmt_type</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_by_type"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_by_type" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to a given statement type.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>stmt_type</strong> (<a class="reference internal" href="../statements.html#indra.statements.Statement" title="indra.statements.Statement"><em>indra.statements.Statement</em></a>) &#8211; The class of the statement type to filter for.
Example: indra.statements.Modification</li>
<li><strong>invert</strong> (<em>Optional[bool]</em>) &#8211; If True, the statements that are not of the given type
are returned. Default: False</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_direct">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_direct</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_direct"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_direct" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to statements that are direct interactions</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_enzyme_kinase">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_enzyme_kinase</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_enzyme_kinase"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_enzyme_kinase" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter Phosphorylations to ones where the enzyme is a known kinase.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_evidence_source">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_evidence_source</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>source_apis</em>, <em>policy=u'one'</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_evidence_source"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_evidence_source" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to statements that have evidence from a given set of sources.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>source_apis</strong> (<em>list[str]</em>) &#8211; A list of sources to filter for. Examples: biopax, bel, reach</li>
<li><strong>policy</strong> (<em>Optional[str]</em>) &#8211; If &#8216;one&#8217;, a statement that hase evidence from any of the sources is
kept. If &#8216;all&#8217;, only those statements are kept which have evidence
from all the input sources specified in source_apis.
If &#8216;none&#8217;, only those statements are kept that don&#8217;t have evidence
from any of the sources specified in source_apis.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_gene_list">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_gene_list</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>gene_list</em>, <em>policy</em>, <em>allow_families=False</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_gene_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_gene_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Return statements that contain genes given in a list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>gene_list</strong> (<em>list[str]</em>) &#8211; A list of gene symbols to filter for.</li>
<li><strong>policy</strong> (<em>str</em>) &#8211; The policy to apply when filtering for the list of genes. &#8220;one&#8221;: keep
statements that contain at least one of the list of genes and
possibly others not in the list &#8220;all&#8221;: keep statements that only
contain genes given in the list</li>
<li><strong>allow_families</strong> (<em>Optional[bool]</em>) &#8211; Will include statements involving Bioentities families containing one
of the genes in the gene list. Default: False</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_genes_only">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_genes_only</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_genes_only"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_genes_only" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to statements containing genes only.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>specific_only</strong> (<em>Optional[bool]</em>) &#8211; If True, only elementary genes/proteins will be kept and families
will be filtered out. If False, families are also included in the
output. Default: False</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_grounded_only">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_grounded_only</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_grounded_only"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_grounded_only" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to statements that have grounded agents.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_human_only">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_human_only</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_human_only"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_human_only" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter out statements that are not grounded to human genes.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_inconsequential_acts">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_inconsequential_acts</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>whitelist=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_inconsequential_acts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_inconsequential_acts" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter out Activations that modify inconsequential activities</p>
<p>Inconsequential here means that the site is not mentioned / tested
in any other statement. In some cases specific activity types should be
preserved, for instance, to be used as readouts in a model.
In this case, the given activities can be passed in a whitelist.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>whitelist</strong> (<em>Optional[dict]</em>) &#8211; A whitelist containing agent activity types which  should be preserved
even if no other statement refers to them.
The whitelist parameter is a dictionary in which
the key is a gene name and the value is a list of activity types.
Example: whitelist = {&#8216;MAP2K1&#8217;: [&#8216;kinase&#8217;]}</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_inconsequential_mods">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_inconsequential_mods</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>whitelist=None</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_inconsequential_mods"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_inconsequential_mods" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter out Modifications that modify inconsequential sites</p>
<p>Inconsequential here means that the site is not mentioned / tested
in any other statement. In some cases specific sites should be
preserved, for instance, to be used as readouts in a model.
In this case, the given sites can be passed in a whitelist.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>whitelist</strong> (<em>Optional[dict]</em>) &#8211; A whitelist containing agent modification sites whose
modifications should be preserved even if no other statement
refers to them. The whitelist parameter is a dictionary in which
the key is a gene name and the value is a list of tuples of
(modification_type, residue, position). Example:
whitelist = {&#8216;MAP2K1&#8217;: [(&#8216;phosphorylation&#8217;, &#8216;S&#8217;, &#8216;222&#8217;)]}</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_mod_nokinase">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_mod_nokinase</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_mod_nokinase"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_mod_nokinase" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter non-phospho Modifications to ones with a non-kinase enzyme.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_mutation_status">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_mutation_status</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>mutations</em>, <em>deletions</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_mutation_status"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_mutation_status" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter statements based on existing mutations/deletions</p>
<p>This filter helps to contextualize a set of statements to a given
cell type. Given a list of deleted genes, it removes statements that refer
to these genes. It also takes a list of mutations and removes statements
that refer to mutations not relevant for the given context.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>mutations</strong> (<em>dict</em>) &#8211; A dictionary whose keys are gene names, and the values are lists of
tuples of the form (residue_from, position, residue_to).
Example: mutations = {&#8216;BRAF&#8217;: [(&#8216;V&#8217;, &#8216;600&#8217;, &#8216;E&#8217;)]}</li>
<li><strong>deletions</strong> (<em>list</em>) &#8211; A list of gene names that are deleted.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_no_hypothesis">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_no_hypothesis</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_no_hypothesis"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_no_hypothesis" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to statements that are not marked as hypothesis in epistemics.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_top_level">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_top_level</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_top_level"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_top_level" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to statements that are at the top-level of the hierarchy.</p>
<p>Here top-level statements correspond to most specific ones.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_transcription_factor">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_transcription_factor</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_transcription_factor"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_transcription_factor" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter out RegulateAmounts where subject is not a transcription factor.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.filter_uuid_list">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">filter_uuid_list</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>uuids</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#filter_uuid_list"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.filter_uuid_list" title="Permalink to this definition">¶</a></dt>
<dd><p>Filter to Statements corresponding to given UUIDs</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to filter.</li>
<li><strong>uuids</strong> (<em>list[str]</em>) &#8211; A list of UUIDs to filter for.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of filtered statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.load_statements">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">load_statements</code><span class="sig-paren">(</span><em>fname</em>, <em>as_dict=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#load_statements"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.load_statements" title="Permalink to this definition">¶</a></dt>
<dd><p>Load statements from a pickle file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>fname</strong> (<em>str</em>) &#8211; The name of the pickle file to load statements from.</li>
<li><strong>as_dict</strong> (<em>Optional[bool]</em>) &#8211; If True and the pickle file contains a dictionary of statements, it
is returned as a dictionary. If False, the statements are always
returned in a list. Default: False</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts</strong> &#8211; A list or dict of statements that were loaded.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.map_grounding">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">map_grounding</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#map_grounding"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.map_grounding" title="Permalink to this definition">¶</a></dt>
<dd><p>Map grounding using the GroundingMapper.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to map.</li>
<li><strong>do_rename</strong> (<em>Optional[bool]</em>) &#8211; If True, Agents are renamed based on their mapped grounding.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of mapped statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.map_sequence">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">map_sequence</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#map_sequence"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.map_sequence" title="Permalink to this definition">¶</a></dt>
<dd><p>Map sequences using the SiteMapper.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to map.</li>
<li><strong>do_methionine_offset</strong> (<em>boolean</em>) &#8211; Whether to check for off-by-one errors in site position (possibly)
attributable to site numbering from mature proteins after
cleavage of the initial methionine. If True, checks the reference
sequence for a known modification at 1 site position greater
than the given one; if there exists such a site, creates the
mapping. Default is True.</li>
<li><strong>do_orthology_mapping</strong> (<em>boolean</em>) &#8211; Whether to check sequence positions for known modification sites
in mouse or rat sequences (based on PhosphoSitePlus data). If a
mouse/rat site is found that is linked to a site in the human
reference sequence, a mapping is created. Default is True.</li>
<li><strong>do_isoform_mapping</strong> (<em>boolean</em>) &#8211; Whether to check sequence positions for known modifications
in other human isoforms of the protein (based on PhosphoSitePlus
data). If a site is found that is linked to a site in the human
reference sequence, a mapping is created. Default is True.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of mapped statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.reduce_activities">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">reduce_activities</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#reduce_activities"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.reduce_activities" title="Permalink to this definition">¶</a></dt>
<dd><p>Reduce the activity types in a list of statements</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to reduce activity types in.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of reduced activity statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.run_preassembly">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">run_preassembly</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#run_preassembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.run_preassembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Run preassembly on a list of statements.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements to preassemble.</li>
<li><strong>return_toplevel</strong> (<em>Optional[bool]</em>) &#8211; If True, only the top-level statements are returned. If False,
all statements are returned irrespective of level of specificity.
Default: True</li>
<li><strong>poolsize</strong> (<em>Optional[int]</em>) &#8211; The number of worker processes to use to parallelize the
comparisons performed by the function. If None (default), no
parallelization is performed. NOTE: Parallelization is only
available on Python 3.4 and above.</li>
<li><strong>size_cutoff</strong> (<em>Optional[int]</em>) &#8211; Groups with size_cutoff or more statements are sent to worker
processes, while smaller groups are compared in the parent process.
Default value is 100. Not relevant when parallelization is not
used.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
<li><strong>save_unique</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the unique statements into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of preassembled top-level statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.run_preassembly_duplicate">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">run_preassembly_duplicate</code><span class="sig-paren">(</span><em>preassembler</em>, <em>beliefengine</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#run_preassembly_duplicate"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.run_preassembly_duplicate" title="Permalink to this definition">¶</a></dt>
<dd><p>Run deduplication stage of preassembly on a list of statements.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>preassembler</strong> (<a class="reference internal" href="../preassembler/index.html#indra.preassembler.Preassembler" title="indra.preassembler.Preassembler"><em>indra.preassembler.Preassembler</em></a>) &#8211; A Preassembler instance</li>
<li><strong>beliefengine</strong> (<a class="reference internal" href="../belief/index.html#indra.belief.BeliefEngine" title="indra.belief.BeliefEngine"><em>indra.belief.BeliefEngine</em></a>) &#8211; A BeliefEngine instance</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of unique statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.run_preassembly_related">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">run_preassembly_related</code><span class="sig-paren">(</span><em>preassembler</em>, <em>beliefengine</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#run_preassembly_related"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.run_preassembly_related" title="Permalink to this definition">¶</a></dt>
<dd><p>Run related stage of preassembly on a list of statements.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>preassembler</strong> (<a class="reference internal" href="../preassembler/index.html#indra.preassembler.Preassembler" title="indra.preassembler.Preassembler"><em>indra.preassembler.Preassembler</em></a>) &#8211; A Preassembler instance which already has a set of unique statements
internally.</li>
<li><strong>beliefengine</strong> (<a class="reference internal" href="../belief/index.html#indra.belief.BeliefEngine" title="indra.belief.BeliefEngine"><em>indra.belief.BeliefEngine</em></a>) &#8211; A BeliefEngine instance</li>
<li><strong>return_toplevel</strong> (<em>Optional[bool]</em>) &#8211; If True, only the top-level statements are returned. If False,
all statements are returned irrespective of level of specificity.
Default: True</li>
<li><strong>poolsize</strong> (<em>Optional[int]</em>) &#8211; The number of worker processes to use to parallelize the
comparisons performed by the function. If None (default), no
parallelization is performed. NOTE: Parallelization is only
available on Python 3.4 and above.</li>
<li><strong>size_cutoff</strong> (<em>Optional[int]</em>) &#8211; Groups with size_cutoff or more statements are sent to worker
processes, while smaller groups are compared in the parent process.
Default value is 100. Not relevant when parallelization is not
used.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of preassembled top-level statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="function">
<dt id="indra.tools.assemble_corpus.strip_agent_context">
<code class="descclassname">indra.tools.assemble_corpus.</code><code class="descname">strip_agent_context</code><span class="sig-paren">(</span><em>stmts_in</em>, <em>**kwargs</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/assemble_corpus.html#strip_agent_context"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.assemble_corpus.strip_agent_context" title="Permalink to this definition">¶</a></dt>
<dd><p>Strip any context on agents within each statement.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts_in</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of statements whose agent context should be stripped.</li>
<li><strong>save</strong> (<em>Optional[str]</em>) &#8211; The name of a pickle file to save the results (stmts_out) into.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>stmts_out</strong> &#8211; A list of stripped statements.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list[indra.statements.Statement]</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-indra.tools.gene_network">
<span id="build-a-network-from-a-gene-list-indra-tools-gene-network"></span><h2>Build a network from a gene list (<a class="reference internal" href="#module-indra.tools.gene_network" title="indra.tools.gene_network"><code class="xref py py-mod docutils literal"><span class="pre">indra.tools.gene_network</span></code></a>)<a class="headerlink" href="#module-indra.tools.gene_network" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="indra.tools.gene_network.GeneNetwork">
<em class="property">class </em><code class="descclassname">indra.tools.gene_network.</code><code class="descname">GeneNetwork</code><span class="sig-paren">(</span><em>gene_list</em>, <em>basename=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/gene_network.html#GeneNetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Build a set of INDRA statements for a given gene list from databases.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>gene_list</strong> (<em>string</em>) &#8211; List of gene names.</li>
<li><strong>basename</strong> (<em>string or None (default)</em>) &#8211; Filename prefix to be used for caching of intermediates (Biopax OWL
file, pickled statement lists, etc.). If None, no results are cached
and no cached files are used.</li>
</ul>
</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="indra.tools.gene_network.GeneNetwork.gene_list">
<code class="descname">gene_list</code><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork.gene_list" title="Permalink to this definition">¶</a></dt>
<dd><p><em>string</em> &#8211; List of gene names</p>
</dd></dl>

<dl class="attribute">
<dt id="indra.tools.gene_network.GeneNetwork.basename">
<code class="descname">basename</code><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork.basename" title="Permalink to this definition">¶</a></dt>
<dd><p><em>string or None</em> &#8211; Filename prefix for cached intermediates, or None if no cached used.</p>
</dd></dl>

<dl class="attribute">
<dt id="indra.tools.gene_network.GeneNetwork.results">
<code class="descname">results</code><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork.results" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dict</em> &#8211; Dict containing results of preassembly (see return type for
<a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.run_preassembly" title="indra.tools.gene_network.GeneNetwork.run_preassembly"><code class="xref py py-meth docutils literal"><span class="pre">run_preassembly()</span></code></a>.</p>
</dd></dl>

<dl class="method">
<dt id="indra.tools.gene_network.GeneNetwork.get_bel_stmts">
<code class="descname">get_bel_stmts</code><span class="sig-paren">(</span><em>filter=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/gene_network.html#GeneNetwork.get_bel_stmts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork.get_bel_stmts" title="Permalink to this definition">¶</a></dt>
<dd><p>Get relevant statements from the BEL large corpus.</p>
<p>Performs a series of neighborhood queries and then takes the union of
all the statements. Because the query process can take a long time for
large gene lists, the resulting list of statements are cached in a
pickle file with the filename <cite>&lt;basename&gt;_bel_stmts.pkl</cite>.  If the
pickle file is present, it is used by default; if not present, the
queries are performed and the results are cached.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filter</strong> (<em>bool</em>) &#8211; If True, includes only those statements that exclusively mention
genes in <a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.gene_list" title="indra.tools.gene_network.GeneNetwork.gene_list"><code class="xref py py-attr docutils literal"><span class="pre">gene_list</span></code></a>. Default is False. Note that the
full (unfiltered) set of statements are cached.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">List of INDRA statements extracted from the BEL large corpus.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of <a class="reference internal" href="../statements.html#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.gene_network.GeneNetwork.get_biopax_stmts">
<code class="descname">get_biopax_stmts</code><span class="sig-paren">(</span><em>filter=False</em>, <em>query=u'pathsbetween'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/gene_network.html#GeneNetwork.get_biopax_stmts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork.get_biopax_stmts" title="Permalink to this definition">¶</a></dt>
<dd><p>Get relevant statements from Pathway Commons.</p>
<p>Performs a &#8220;paths between&#8221; query for the genes in <a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.gene_list" title="indra.tools.gene_network.GeneNetwork.gene_list"><code class="xref py py-attr docutils literal"><span class="pre">gene_list</span></code></a>
and uses the results to build statements. This function caches two
files: the list of statements built from the query, which is cached in
<cite>&lt;basename&gt;_biopax_stmts.pkl</cite>, and the OWL file returned by the Pathway
Commons Web API, which is cached in <cite>&lt;basename&gt;_pc_pathsbetween.owl</cite>.
If these cached files are found, then the results are returned based
on the cached file and Pathway Commons is not queried again.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>filter</strong> (<em>bool</em>) &#8211; If True, includes only those statements that exclusively mention
genes in <a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.gene_list" title="indra.tools.gene_network.GeneNetwork.gene_list"><code class="xref py py-attr docutils literal"><span class="pre">gene_list</span></code></a>. Default is False.</li>
<li><strong>query</strong> (<em>str</em>) &#8211; Defined what type of query is executed. The two options are
&#8216;pathsbetween&#8217; which finds paths between the given list of genes
and only works if more than 1 gene is given, and &#8216;neighborhood&#8217;
which searches the immediate neighborhood of each given gene.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first">List of INDRA statements extracted from Pathway Commons.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">list of <a class="reference internal" href="../statements.html#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.gene_network.GeneNetwork.get_statements">
<code class="descname">get_statements</code><span class="sig-paren">(</span><em>filter=False</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/gene_network.html#GeneNetwork.get_statements"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork.get_statements" title="Permalink to this definition">¶</a></dt>
<dd><p>Return the combined list of statements from BEL and Pathway Commons.</p>
<p>Internally calls <a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.get_biopax_stmts" title="indra.tools.gene_network.GeneNetwork.get_biopax_stmts"><code class="xref py py-meth docutils literal"><span class="pre">get_biopax_stmts()</span></code></a> and
<a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.get_bel_stmts" title="indra.tools.gene_network.GeneNetwork.get_bel_stmts"><code class="xref py py-meth docutils literal"><span class="pre">get_bel_stmts()</span></code></a>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filter</strong> (<em>bool</em>) &#8211; If True, includes only those statements that exclusively mention
genes in <a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.gene_list" title="indra.tools.gene_network.GeneNetwork.gene_list"><code class="xref py py-attr docutils literal"><span class="pre">gene_list</span></code></a>. Default is False.</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body">List of INDRA statements extracted the BEL large corpus and Pathway
Commons.</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body">list of <a class="reference internal" href="../statements.html#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a></td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.gene_network.GeneNetwork.run_preassembly">
<code class="descname">run_preassembly</code><span class="sig-paren">(</span><em>stmts</em>, <em>print_summary=True</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/gene_network.html#GeneNetwork.run_preassembly"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.gene_network.GeneNetwork.run_preassembly" title="Permalink to this definition">¶</a></dt>
<dd><p>Run complete preassembly procedure on the given statements.</p>
<p>Results are returned as a dict and stored in the attribute
<a class="reference internal" href="#indra.tools.gene_network.GeneNetwork.results" title="indra.tools.gene_network.GeneNetwork.results"><code class="xref py py-attr docutils literal"><span class="pre">results</span></code></a>. They are also saved in the pickle file
<cite>&lt;basename&gt;_results.pkl</cite>.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>stmts</strong> (list of <a class="reference internal" href="../statements.html#indra.statements.Statement" title="indra.statements.Statement"><code class="xref py py-class docutils literal"><span class="pre">indra.statements.Statement</span></code></a>) &#8211; Statements to preassemble.</li>
<li><strong>print_summary</strong> (<em>bool</em>) &#8211; If True (default), prints a summary of the preassembly process to
the console.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><p>A dict containing the following entries:</p>
<ul class="simple">
<li><cite>raw</cite>: the starting set of statements before preassembly.</li>
<li><cite>duplicates1</cite>: statements after initial de-duplication.</li>
<li><cite>valid</cite>: statements found to have valid modification sites.</li>
<li><cite>mapped</cite>: mapped statements (list of
<a class="reference internal" href="../preassembler/index.html#indra.preassembler.sitemapper.MappedStatement" title="indra.preassembler.sitemapper.MappedStatement"><code class="xref py py-class docutils literal"><span class="pre">indra.preassembler.sitemapper.MappedStatement</span></code></a>).</li>
<li><cite>mapped_stmts</cite>: combined list of valid statements and statements
after mapping.</li>
<li><cite>duplicates2</cite>: statements resulting from de-duplication of the
statements in <cite>mapped_stmts</cite>.</li>
<li><cite>related2</cite>: top-level statements after combining the statements
in <cite>duplicates2</cite>.</li>
</ul>
</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">dict</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-indra.tools.executable_subnetwork">
<span id="build-an-executable-model-from-a-fragment-of-a-large-network-indra-tools-executable-subnetwork"></span><h2>Build an executable model from a fragment of a large network (<a class="reference internal" href="#module-indra.tools.executable_subnetwork" title="indra.tools.executable_subnetwork"><code class="xref py py-mod docutils literal"><span class="pre">indra.tools.executable_subnetwork</span></code></a>)<a class="headerlink" href="#module-indra.tools.executable_subnetwork" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="indra.tools.executable_subnetwork.get_subnetwork">
<code class="descclassname">indra.tools.executable_subnetwork.</code><code class="descname">get_subnetwork</code><span class="sig-paren">(</span><em>statements</em>, <em>nodes</em>, <em>relevance_network=None</em>, <em>relevance_node_lim=10</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/executable_subnetwork.html#get_subnetwork"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.executable_subnetwork.get_subnetwork" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a PySB model based on a subset of given INDRA Statements.</p>
<p>Statements are first filtered for nodes in the given list and other nodes
are optionally added based on relevance in a given network. The filtered
statements are then assembled into an executable model using INDRA&#8217;s
PySB Assembler.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first simple">
<li><strong>statements</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of INDRA Statements to extract a subnetwork from.</li>
<li><strong>nodes</strong> (<em>list[str]</em>) &#8211; The names of the nodes to extract the subnetwork for.</li>
<li><strong>relevance_network</strong> (<em>Optional[str]</em>) &#8211; The UUID of the NDEx network in which nodes relevant to the given
nodes are found.</li>
<li><strong>relevance_node_lim</strong> (<em>Optional[int]</em>) &#8211; The maximal number of additional nodes to add to the subnetwork
based on relevance.</li>
</ul>
</td>
</tr>
<tr class="field-even field"><th class="field-name">Returns:</th><td class="field-body"><p class="first"><strong>model</strong> &#8211; A PySB model object assembled using INDRA&#8217;s PySB Assembler from
the INDRA Statements corresponding to the subnetwork.</p>
</td>
</tr>
<tr class="field-odd field"><th class="field-name">Return type:</th><td class="field-body"><p class="first last">pysb.Model</p>
</td>
</tr>
</tbody>
</table>
</dd></dl>

</div>
<div class="section" id="module-indra.tools.incremental_model">
<span id="build-a-model-incrementally-over-time-indra-tools-incremental-model"></span><h2>Build a model incrementally over time (<a class="reference internal" href="#module-indra.tools.incremental_model" title="indra.tools.incremental_model"><code class="xref py py-mod docutils literal"><span class="pre">indra.tools.incremental_model</span></code></a>)<a class="headerlink" href="#module-indra.tools.incremental_model" title="Permalink to this headline">¶</a></h2>
<dl class="class">
<dt id="indra.tools.incremental_model.IncrementalModel">
<em class="property">class </em><code class="descclassname">indra.tools.incremental_model.</code><code class="descname">IncrementalModel</code><span class="sig-paren">(</span><em>model_fname=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel" title="Permalink to this definition">¶</a></dt>
<dd><p>Assemble a model incrementally by iteratively adding new Statements.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>model_fname</strong> (<em>Optional[str]</em>) &#8211; The name of the pickle file in which a set of INDRA Statements are
stored in a dict keyed by PubMed IDs. This is the state of an
IncrementalModel that is loaded upon instantiation.</td>
</tr>
</tbody>
</table>
<dl class="attribute">
<dt id="indra.tools.incremental_model.IncrementalModel.stmts">
<code class="descname">stmts</code><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.stmts" title="Permalink to this definition">¶</a></dt>
<dd><p><em>dict[str, list[indra.statements.Statement]]</em> &#8211; A dictionary of INDRA Statements keyed by PMIDs that stores the current
state of the IncrementalModel.</p>
</dd></dl>

<dl class="attribute">
<dt id="indra.tools.incremental_model.IncrementalModel.assembled_stmts">
<code class="descname">assembled_stmts</code><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.assembled_stmts" title="Permalink to this definition">¶</a></dt>
<dd><p><em>list[indra.statements.Statement]</em> &#8211; A list of INDRA Statements after assembly.</p>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.add_statements">
<code class="descname">add_statements</code><span class="sig-paren">(</span><em>pmid</em>, <em>stmts</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.add_statements"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.add_statements" title="Permalink to this definition">¶</a></dt>
<dd><p>Add INDRA Statements to the incremental model indexed by PMID.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><ul class="first last simple">
<li><strong>pmid</strong> (<em>str</em>) &#8211; The PMID of the paper from which statements were extracted.</li>
<li><strong>stmts</strong> (<em>list[indra.statements.Statement]</em>) &#8211; A list of INDRA Statements to be added to the model.</li>
</ul>
</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.get_model_agents">
<code class="descname">get_model_agents</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.get_model_agents"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.get_model_agents" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of all Agents from all Statements.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>agents</strong> &#8211; A list of Agents that are in the model.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list[indra.statements.Agent]</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.get_statements">
<code class="descname">get_statements</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.get_statements"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.get_statements" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of all Statements in a single list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>stmts</strong> &#8211; A list of all the INDRA Statements in the model.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list[indra.statements.Statement]</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.get_statements_noprior">
<code class="descname">get_statements_noprior</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.get_statements_noprior"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.get_statements_noprior" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of all non-prior Statements in a single list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>stmts</strong> &#8211; A list of all the INDRA Statements in the model (excluding
the prior).</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list[indra.statements.Statement]</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.get_statements_prior">
<code class="descname">get_statements_prior</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.get_statements_prior"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.get_statements_prior" title="Permalink to this definition">¶</a></dt>
<dd><p>Return a list of all prior Statements in a single list.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Returns:</th><td class="field-body"><strong>stmts</strong> &#8211; A list of all the INDRA Statements in the prior.</td>
</tr>
<tr class="field-even field"><th class="field-name">Return type:</th><td class="field-body">list[indra.statements.Statement]</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.load_prior">
<code class="descname">load_prior</code><span class="sig-paren">(</span><em>prior_fname</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.load_prior"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.load_prior" title="Permalink to this definition">¶</a></dt>
<dd><p>Load a set of prior statements from a pickle file.</p>
<p>The prior statements have a special key in the stmts dictionary
called &#8220;prior&#8221;.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>prior_fname</strong> (<em>str</em>) &#8211; The name of the pickle file containing the prior Statements.</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.preassemble">
<code class="descname">preassemble</code><span class="sig-paren">(</span><em>filters=None</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.preassemble"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.preassemble" title="Permalink to this definition">¶</a></dt>
<dd><p>Preassemble the Statements collected in the model.</p>
<p>Use INDRA&#8217;s GroundingMapper, Preassembler and BeliefEngine
on the IncrementalModel and save the unique statements and
the top level statements in class attributes.</p>
<p>Currently the following filter options are implemented:
- grounding: require that all Agents in statements are grounded
- human_only: require that all proteins are human proteins
- prior_one: require that at least one Agent is in the prior model
- prior_all: require that all Agents are in the prior model</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>filters</strong> (<em>Optional[list[str]]</em>) &#8211; A list of filter options to apply when choosing the statements.
See description above for more details. Default: None</td>
</tr>
</tbody>
</table>
</dd></dl>

<dl class="method">
<dt id="indra.tools.incremental_model.IncrementalModel.save">
<code class="descname">save</code><span class="sig-paren">(</span><em>model_fname=u'model.pkl'</em><span class="sig-paren">)</span><a class="reference internal" href="../../_modules/indra/tools/incremental_model.html#IncrementalModel.save"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#indra.tools.incremental_model.IncrementalModel.save" title="Permalink to this definition">¶</a></dt>
<dd><p>Save the state of the IncrementalModel in a pickle file.</p>
<table class="docutils field-list" frame="void" rules="none">
<col class="field-name" />
<col class="field-body" />
<tbody valign="top">
<tr class="field-odd field"><th class="field-name">Parameters:</th><td class="field-body"><strong>model_fname</strong> (<em>Optional[str]</em>) &#8211; The name of the pickle file to save the state of the
IncrementalModel in. Default: model.pkl</td>
</tr>
</tbody>
</table>
</dd></dl>

</dd></dl>

</div>
</div>


          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../../tutorials/index.html" class="btn btn-neutral float-right" title="Tutorials" accesskey="n">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="../explanation/index.html" class="btn btn-neutral" title="Explanation (indra.explanation)" accesskey="p"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, B. M. Gyori, J. A. Bachman.
    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'../../',
            VERSION:'1.5.0',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true
        };
    </script>
      <script type="text/javascript" src="../../_static/jquery.js"></script>
      <script type="text/javascript" src="../../_static/underscore.js"></script>
      <script type="text/javascript" src="../../_static/doctools.js"></script>

  

  
  
    <script type="text/javascript" src="../../_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>